// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/keywords/utils/curry.js
var _CurriedTypeToReadabl;

import { ASTv2, generateSyntaxError } from '@glimmer/syntax';
import { Err, Ok, Result } from '../../../../shared/result';
import * as mir from '../../../2-encoding/mir';
import { VISIT_EXPRS } from '../../visitors/expressions';
var CurriedTypeToReadableType = (_CurriedTypeToReadabl = {}, _CurriedTypeToReadabl[0
/* Component */
] = 'component', _CurriedTypeToReadabl[1
/* Helper */
] = 'helper', _CurriedTypeToReadabl[2
/* Modifier */
] = 'modifier', _CurriedTypeToReadabl);
export function assertCurryKeyword(curriedType) {
  return function (node, state) {
    var readableType = CurriedTypeToReadableType[curriedType];
    var stringsAllowed = curriedType === 0
    /* Component */
    ;
    var args = node.args;
    var definition = args.nth(0);

    if (definition === null) {
      return Err(generateSyntaxError("(" + readableType + ") requires a " + readableType + " definition or identifier as its first positional parameter, did not receive any parameters.", args.loc));
    }

    if (definition.type === 'Literal') {
      if (stringsAllowed && state.isStrict) {
        return Err(generateSyntaxError("(" + readableType + ") cannot resolve string values in strict mode templates", node.loc));
      } else if (!stringsAllowed) {
        return Err(generateSyntaxError("(" + readableType + ") cannot resolve string values, you must pass a " + readableType + " definition directly", node.loc));
      }
    }

    args = new ASTv2.Args({
      positional: new ASTv2.PositionalArguments({
        exprs: args.positional.exprs.slice(1),
        loc: args.positional.loc
      }),
      named: args.named,
      loc: args.loc
    });
    return Ok({
      definition: definition,
      args: args
    });
  };
}

function translateCurryKeyword(curriedType) {
  return function (_ref, _ref2) {
    var node = _ref.node,
        state = _ref.state;
    var definition = _ref2.definition,
        args = _ref2.args;
    var definitionResult = VISIT_EXPRS.visit(definition, state);
    var argsResult = VISIT_EXPRS.Args(args, state);
    return Result.all(definitionResult, argsResult).mapOk(function (_ref3) {
      var definition = _ref3[0],
          args = _ref3[1];
      return new mir.Curry({
        loc: node.loc,
        curriedType: curriedType,
        definition: definition,
        args: args
      });
    });
  };
}

export function curryKeyword(curriedType) {
  return {
    assert: assertCurryKeyword(curriedType),
    translate: translateCurryKeyword(curriedType)
  };
}
