// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/keywords/utils/dynamic-vars.js
import { generateSyntaxError } from '@glimmer/syntax';
import { Err, Ok } from '../../../../shared/result';
import * as mir from '../../../2-encoding/mir';
import { VISIT_EXPRS } from '../../visitors/expressions';

function assertGetDynamicVarKeyword(node) {
  var call = node.type === 'AppendContent' ? node.value : node;
  var named = call.type === 'Call' ? call.args.named : null;
  var positionals = call.type === 'Call' ? call.args.positional : null;

  if (named && !named.isEmpty()) {
    return Err(generateSyntaxError("(-get-dynamic-vars) does not take any named arguments", node.loc));
  }

  var varName = positionals === null || positionals === void 0 ? void 0 : positionals.nth(0);

  if (!varName) {
    return Err(generateSyntaxError("(-get-dynamic-vars) requires a var name to get", node.loc));
  }

  if (positionals && positionals.size > 1) {
    return Err(generateSyntaxError("(-get-dynamic-vars) only receives one positional arg", node.loc));
  }

  return Ok(varName);
}

function translateGetDynamicVarKeyword(_ref, name) {
  var node = _ref.node,
      state = _ref.state;
  return VISIT_EXPRS.visit(name, state).mapOk(function (name) {
    return new mir.GetDynamicVar({
      name: name,
      loc: node.loc
    });
  });
}

export var getDynamicVarKeyword = {
  assert: assertGetDynamicVarKeyword,
  translate: translateGetDynamicVarKeyword
};
