// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/keywords/utils/has-block.js
import { ASTv2, generateSyntaxError, SourceSlice } from '@glimmer/syntax';
import { Err, Ok } from '../../../../shared/result';
import * as mir from '../../../2-encoding/mir';

function assertHasBlockKeyword(type) {
  return function (node) {
    var call = node.type === 'AppendContent' ? node.value : node;
    var named = call.type === 'Call' ? call.args.named : null;
    var positionals = call.type === 'Call' ? call.args.positional : null;

    if (named && !named.isEmpty()) {
      return Err(generateSyntaxError("(" + type + ") does not take any named arguments", call.loc));
    }

    if (!positionals || positionals.isEmpty()) {
      return Ok(SourceSlice.synthetic('default'));
    } else if (positionals.exprs.length === 1) {
      var positional = positionals.exprs[0];

      if (ASTv2.isLiteral(positional, 'string')) {
        return Ok(positional.toSlice());
      } else {
        return Err(generateSyntaxError("(" + type + ") can only receive a string literal as its first argument", call.loc));
      }
    } else {
      return Err(generateSyntaxError("(" + type + ") only takes a single positional argument", call.loc));
    }
  };
}

function translateHasBlockKeyword(type) {
  return function (_ref, target) {
    var node = _ref.node,
        scope = _ref.state.scope;
    var block = type === 'has-block' ? new mir.HasBlock({
      loc: node.loc,
      target: target,
      symbol: scope.allocateBlock(target.chars)
    }) : new mir.HasBlockParams({
      loc: node.loc,
      target: target,
      symbol: scope.allocateBlock(target.chars)
    });
    return Ok(block);
  };
}

export function hasBlockKeyword(type) {
  return {
    assert: assertHasBlockKeyword(type),
    translate: translateHasBlockKeyword(type)
  };
}
