// @glimmer/opcode-compiler/dist/modules/es5/lib/opcode-builder/operands.js
import { assert, isSmallInt } from '@glimmer/util';
export function labelOperand(value) {
  return {
    type: 1
    /* Label */
    ,
    value: value
  };
}
export function evalSymbolsOperand() {
  return {
    type: 3
    /* EvalSymbols */
    ,
    value: undefined
  };
}
export function isStrictMode() {
  return {
    type: 2
    /* IsStrictMode */
    ,
    value: undefined
  };
}
export function blockOperand(value) {
  return {
    type: 4
    /* Block */
    ,
    value: value
  };
}
export function stdlibOperand(value) {
  return {
    type: 5
    /* StdLib */
    ,
    value: value
  };
}
export function nonSmallIntOperand(value) {
  false && assert(!isSmallInt(value), 'Attempted to make a operand for an int that was not a small int, you should encode this as an immediate');
  return {
    type: 6
    /* NonSmallInt */
    ,
    value: value
  };
}
export function symbolTableOperand(value) {
  return {
    type: 7
    /* SymbolTable */
    ,
    value: value
  };
}
export function layoutOperand(value) {
  return {
    type: 8
    /* Layout */
    ,
    value: value
  };
}
