// @glimmer/opcode-compiler/dist/modules/es5/lib/opcode-builder/stdlib.js
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

export var StdLib = /*#__PURE__*/function () {
  function StdLib(main, trustingGuardedAppend, cautiousGuardedAppend, trustingNonDynamicAppend, cautiousNonDynamicAppend) {
    this.main = main;
    this.trustingGuardedAppend = trustingGuardedAppend;
    this.cautiousGuardedAppend = cautiousGuardedAppend;
    this.trustingNonDynamicAppend = trustingNonDynamicAppend;
    this.cautiousNonDynamicAppend = cautiousNonDynamicAppend;
  }

  var _proto = StdLib.prototype;

  _proto.getAppend = function getAppend(trusting) {
    return trusting ? this.trustingGuardedAppend : this.cautiousGuardedAppend;
  };

  _createClass(StdLib, [{
    key: 'trusting-append',
    get: function get() {
      return this.trustingGuardedAppend;
    }
  }, {
    key: 'cautious-append',
    get: function get() {
      return this.cautiousGuardedAppend;
    }
  }, {
    key: 'trusting-non-dynamic-append',
    get: function get() {
      return this.trustingNonDynamicAppend;
    }
  }, {
    key: 'cautious-non-dynamic-append',
    get: function get() {
      return this.cautiousNonDynamicAppend;
    }
  }]);

  return StdLib;
}();
