// @glimmer/runtime/dist/modules/es5/lib/dom/helper.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { applySVGInnerHTMLFix } from '../compat/svg-inner-html-fix';
import { applyTextNodeMergingFix } from '../compat/text-node-merging-fix';
import { BLACKLIST_TABLE, DOMOperations } from './operations';
['b', 'big', 'blockquote', 'body', 'br', 'center', 'code', 'dd', 'div', 'dl', 'dt', 'em', 'embed', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'hr', 'i', 'img', 'li', 'listing', 'main', 'meta', 'nobr', 'ol', 'p', 'pre', 'ruby', 's', 'small', 'span', 'strong', 'strike', 'sub', 'sup', 'table', 'tt', 'u', 'ul', 'var'].forEach(function (tag) {
  return BLACKLIST_TABLE[tag] = 1;
});
var WHITESPACE = /[\t-\r \xA0\u1680\u180E\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]/;
var doc = typeof document === 'undefined' ? null : document;
export function isWhitespace(string) {
  return WHITESPACE.test(string);
}
export var DOM;

(function (DOM) {
  var TreeConstruction = /*#__PURE__*/function (_DOMOperations) {
    _inheritsLoose(TreeConstruction, _DOMOperations);

    function TreeConstruction() {
      return _DOMOperations.apply(this, arguments) || this;
    }

    var _proto = TreeConstruction.prototype;

    _proto.createElementNS = function createElementNS(namespace, tag) {
      return this.document.createElementNS(namespace, tag);
    };

    _proto.setAttribute = function setAttribute(element, name, value, namespace) {
      if (namespace === void 0) {
        namespace = null;
      }

      if (namespace) {
        element.setAttributeNS(namespace, name, value);
      } else {
        element.setAttribute(name, value);
      }
    };

    return TreeConstruction;
  }(DOMOperations);

  DOM.TreeConstruction = TreeConstruction;
  var appliedTreeConstruction = TreeConstruction;
  appliedTreeConstruction = applyTextNodeMergingFix(doc, appliedTreeConstruction);
  appliedTreeConstruction = applySVGInnerHTMLFix(doc, appliedTreeConstruction, "http://www.w3.org/2000/svg"
  /* SVG */
  );
  DOM.DOMTreeConstruction = appliedTreeConstruction;
})(DOM || (DOM = {}));

export var DOMChangesImpl = /*#__PURE__*/function (_DOMOperations2) {
  _inheritsLoose(DOMChangesImpl, _DOMOperations2);

  function DOMChangesImpl(document) {
    var _this;

    _this = _DOMOperations2.call(this, document) || this;
    _this.document = document;
    _this.namespace = null;
    return _this;
  }

  var _proto2 = DOMChangesImpl.prototype;

  _proto2.setAttribute = function setAttribute(element, name, value) {
    element.setAttribute(name, value);
  };

  _proto2.removeAttribute = function removeAttribute(element, name) {
    element.removeAttribute(name);
  };

  _proto2.insertAfter = function insertAfter(element, node, reference) {
    this.insertBefore(element, node, reference.nextSibling);
  };

  return DOMChangesImpl;
}(DOMOperations);
var helper = DOMChangesImpl;
helper = applyTextNodeMergingFix(doc, helper);
helper = applySVGInnerHTMLFix(doc, helper, "http://www.w3.org/2000/svg"
/* SVG */
);
export default helper;
export var DOMTreeConstruction = DOM.DOMTreeConstruction;
