// @glimmer/vm/dist/commonjs/es2017/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isMachineOp", {
  enumerable: true,
  get: function () {
    return _opcodes.isMachineOp;
  }
});
Object.defineProperty(exports, "isOp", {
  enumerable: true,
  get: function () {
    return _opcodes.isOp;
  }
});
Object.defineProperty(exports, "SavedRegister", {
  enumerable: true,
  get: function () {
    return _registers.SavedRegister;
  }
});
Object.defineProperty(exports, "TemporaryRegister", {
  enumerable: true,
  get: function () {
    return _registers.TemporaryRegister;
  }
});
Object.defineProperty(exports, "isLowLevelRegister", {
  enumerable: true,
  get: function () {
    return _registers.isLowLevelRegister;
  }
});
Object.defineProperty(exports, "$pc", {
  enumerable: true,
  get: function () {
    return _registers.$pc;
  }
});
Object.defineProperty(exports, "$fp", {
  enumerable: true,
  get: function () {
    return _registers.$fp;
  }
});
Object.defineProperty(exports, "$ra", {
  enumerable: true,
  get: function () {
    return _registers.$ra;
  }
});
Object.defineProperty(exports, "$sp", {
  enumerable: true,
  get: function () {
    return _registers.$sp;
  }
});
Object.defineProperty(exports, "$s0", {
  enumerable: true,
  get: function () {
    return _registers.$s0;
  }
});
Object.defineProperty(exports, "$s1", {
  enumerable: true,
  get: function () {
    return _registers.$s1;
  }
});
Object.defineProperty(exports, "$t0", {
  enumerable: true,
  get: function () {
    return _registers.$t0;
  }
});
Object.defineProperty(exports, "$t1", {
  enumerable: true,
  get: function () {
    return _registers.$t1;
  }
});
Object.defineProperty(exports, "$v0", {
  enumerable: true,
  get: function () {
    return _registers.$v0;
  }
});

var _opcodes = require("./lib/opcodes");

var _registers = require("./lib/registers");
