// @glimmer/manager/dist/commonjs/es2017/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "setInternalHelperManager", {
  enumerable: true,
  get: function () {
    return _index.setInternalHelperManager;
  }
});
Object.defineProperty(exports, "setInternalModifierManager", {
  enumerable: true,
  get: function () {
    return _index.setInternalModifierManager;
  }
});
Object.defineProperty(exports, "setInternalComponentManager", {
  enumerable: true,
  get: function () {
    return _index.setInternalComponentManager;
  }
});
Object.defineProperty(exports, "getInternalHelperManager", {
  enumerable: true,
  get: function () {
    return _index.getInternalHelperManager;
  }
});
Object.defineProperty(exports, "getInternalModifierManager", {
  enumerable: true,
  get: function () {
    return _index.getInternalModifierManager;
  }
});
Object.defineProperty(exports, "getInternalComponentManager", {
  enumerable: true,
  get: function () {
    return _index.getInternalComponentManager;
  }
});
Object.defineProperty(exports, "hasInternalHelperManager", {
  enumerable: true,
  get: function () {
    return _index.hasInternalHelperManager;
  }
});
Object.defineProperty(exports, "hasInternalModifierManager", {
  enumerable: true,
  get: function () {
    return _index.hasInternalModifierManager;
  }
});
Object.defineProperty(exports, "hasInternalComponentManager", {
  enumerable: true,
  get: function () {
    return _index.hasInternalComponentManager;
  }
});
Object.defineProperty(exports, "setHelperManager", {
  enumerable: true,
  get: function () {
    return _index2.setHelperManager;
  }
});
Object.defineProperty(exports, "setModifierManager", {
  enumerable: true,
  get: function () {
    return _index2.setModifierManager;
  }
});
Object.defineProperty(exports, "setComponentManager", {
  enumerable: true,
  get: function () {
    return _index2.setComponentManager;
  }
});
Object.defineProperty(exports, "componentCapabilities", {
  enumerable: true,
  get: function () {
    return _component.componentCapabilities;
  }
});
Object.defineProperty(exports, "CustomComponentManager", {
  enumerable: true,
  get: function () {
    return _component.CustomComponentManager;
  }
});
Object.defineProperty(exports, "modifierCapabilities", {
  enumerable: true,
  get: function () {
    return _modifier.modifierCapabilities;
  }
});
Object.defineProperty(exports, "CustomModifierManager", {
  enumerable: true,
  get: function () {
    return _modifier.CustomModifierManager;
  }
});
Object.defineProperty(exports, "helperCapabilities", {
  enumerable: true,
  get: function () {
    return _helper.helperCapabilities;
  }
});
Object.defineProperty(exports, "hasDestroyable", {
  enumerable: true,
  get: function () {
    return _helper.hasDestroyable;
  }
});
Object.defineProperty(exports, "hasValue", {
  enumerable: true,
  get: function () {
    return _helper.hasValue;
  }
});
Object.defineProperty(exports, "CustomHelperManager", {
  enumerable: true,
  get: function () {
    return _helper.CustomHelperManager;
  }
});
Object.defineProperty(exports, "getComponentTemplate", {
  enumerable: true,
  get: function () {
    return _template.getComponentTemplate;
  }
});
Object.defineProperty(exports, "setComponentTemplate", {
  enumerable: true,
  get: function () {
    return _template.setComponentTemplate;
  }
});
Object.defineProperty(exports, "capabilityFlagsFrom", {
  enumerable: true,
  get: function () {
    return _capabilities.capabilityFlagsFrom;
  }
});
Object.defineProperty(exports, "hasCapability", {
  enumerable: true,
  get: function () {
    return _capabilities.hasCapability;
  }
});
Object.defineProperty(exports, "managerHasCapability", {
  enumerable: true,
  get: function () {
    return _capabilities.managerHasCapability;
  }
});
Object.defineProperty(exports, "getCustomTagFor", {
  enumerable: true,
  get: function () {
    return _argsProxy.getCustomTagFor;
  }
});
Object.defineProperty(exports, "setCustomTagFor", {
  enumerable: true,
  get: function () {
    return _argsProxy.setCustomTagFor;
  }
});

var _index = require("./lib/internal/index");

var _index2 = require("./lib/public/index");

var _component = require("./lib/public/component");

var _modifier = require("./lib/public/modifier");

var _helper = require("./lib/public/helper");

var _template = require("./lib/public/template");

var _capabilities = require("./lib/util/capabilities");

var _argsProxy = require("./lib/util/args-proxy");
