// @glimmer/runtime/dist/commonjs/es2017/lib/dom/normalize.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.normalizeStringValue = normalizeStringValue;
exports.normalizeTrustedValue = normalizeTrustedValue;
exports.shouldCoerce = shouldCoerce;
exports.isEmpty = isEmpty;
exports.isSafeString = isSafeString;
exports.isNode = isNode;
exports.isFragment = isFragment;
exports.isString = isString;

function normalizeStringValue(value) {
  if (isEmpty(value)) {
    return '';
  }

  return String(value);
}

function normalizeTrustedValue(value) {
  if (isEmpty(value)) {
    return '';
  }

  if (isString(value)) {
    return value;
  }

  if (isSafeString(value)) {
    return value.toHTML();
  }

  if (isNode(value)) {
    return value;
  }

  return String(value);
}

function shouldCoerce(value) {
  return isString(value) || isEmpty(value) || typeof value === 'boolean' || typeof value === 'number';
}

function isEmpty(value) {
  return value === null || value === undefined || typeof value.toString !== 'function';
}

function isSafeString(value) {
  return typeof value === 'object' && value !== null && typeof value.toHTML === 'function';
}

function isNode(value) {
  return typeof value === 'object' && value !== null && typeof value.nodeType === 'number';
}

function isFragment(value) {
  return isNode(value) && value.nodeType === 11;
}

function isString(value) {
  return typeof value === 'string';
}
