// @glimmer/opcode-compiler/dist/commonjs/es2017/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  debugCompiler: true,
  compileStatements: true,
  compilable: true,
  invokeStaticBlockWithStack: true,
  invokeStaticBlock: true,
  compileStd: true,
  meta: true,
  StdLib: true,
  templateFactory: true,
  templateCacheCounters: true,
  WrappedBuilder: true,
  EMPTY_BLOCKS: true
};
Object.defineProperty(exports, "debugCompiler", {
  enumerable: true,
  get: function () {
    return _compiler.debugCompiler;
  }
});
Object.defineProperty(exports, "compileStatements", {
  enumerable: true,
  get: function () {
    return _compilableTemplate.compileStatements;
  }
});
Object.defineProperty(exports, "compilable", {
  enumerable: true,
  get: function () {
    return _compilableTemplate.compilable;
  }
});
Object.defineProperty(exports, "invokeStaticBlockWithStack", {
  enumerable: true,
  get: function () {
    return _blocks.InvokeStaticBlockWithStack;
  }
});
Object.defineProperty(exports, "invokeStaticBlock", {
  enumerable: true,
  get: function () {
    return _blocks.InvokeStaticBlock;
  }
});
Object.defineProperty(exports, "compileStd", {
  enumerable: true,
  get: function () {
    return _stdlib.compileStd;
  }
});
Object.defineProperty(exports, "meta", {
  enumerable: true,
  get: function () {
    return _shared.meta;
  }
});
Object.defineProperty(exports, "StdLib", {
  enumerable: true,
  get: function () {
    return _stdlib2.StdLib;
  }
});
Object.defineProperty(exports, "templateFactory", {
  enumerable: true,
  get: function () {
    return _template.default;
  }
});
Object.defineProperty(exports, "templateCacheCounters", {
  enumerable: true,
  get: function () {
    return _template.templateCacheCounters;
  }
});
Object.defineProperty(exports, "WrappedBuilder", {
  enumerable: true,
  get: function () {
    return _wrappedComponent.WrappedBuilder;
  }
});
Object.defineProperty(exports, "EMPTY_BLOCKS", {
  enumerable: true,
  get: function () {
    return _utils.EMPTY_BLOCKS;
  }
});

var _programContext = require("./lib/program-context");

Object.keys(_programContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _programContext[key];
    }
  });
});

var _compiler = require("./lib/compiler");

var _compilableTemplate = require("./lib/compilable-template");

var _context = require("./lib/opcode-builder/context");

Object.keys(_context).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _context[key];
    }
  });
});

var _delegate = require("./lib/opcode-builder/delegate");

Object.keys(_delegate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _delegate[key];
    }
  });
});

var _blocks = require("./lib/opcode-builder/helpers/blocks");

var _stdlib = require("./lib/opcode-builder/helpers/stdlib");

var _shared = require("./lib/opcode-builder/helpers/shared");

var _stdlib2 = require("./lib/opcode-builder/stdlib");

var _template = _interopRequireWildcard(require("./lib/template"));

var _wrappedComponent = require("./lib/wrapped-component");

var _utils = require("./lib/utils");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
