// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/keywords/append.js
import { ASTv2, generateSyntaxError, SourceSpan } from '@glimmer/syntax';
import { Err, Ok, Result } from '../../../shared/result';
import * as mir from '../../2-encoding/mir';
import { VISIT_EXPRS } from '../visitors/expressions';
import { keywords } from './impl';
import { toAppend } from './utils/call-to-append';
import { assertCurryKeyword } from './utils/curry';
import { getDynamicVarKeyword } from './utils/dynamic-vars';
import { hasBlockKeyword } from './utils/has-block';
import { ifUnlessInlineKeyword } from './utils/if-unless';
import { logKeyword } from './utils/log';
export var APPEND_KEYWORDS = keywords('Append').kw('has-block', toAppend(hasBlockKeyword('has-block'))).kw('has-block-params', toAppend(hasBlockKeyword('has-block-params'))).kw('-get-dynamic-var', toAppend(getDynamicVarKeyword)).kw('log', toAppend(logKeyword)).kw('if', toAppend(ifUnlessInlineKeyword('if'))).kw('unless', toAppend(ifUnlessInlineKeyword('unless'))).kw('yield', {
  assert: function assert(node) {
    var args = node.args;

    if (args.named.isEmpty()) {
      return Ok({
        target: SourceSpan.synthetic('default').toSlice(),
        positional: args.positional
      });
    } else {
      var target = args.named.get('to');

      if (args.named.size > 1 || target === null) {
        return Err(generateSyntaxError("yield only takes a single named argument: 'to'", args.named.loc));
      }

      if (ASTv2.isLiteral(target, 'string')) {
        return Ok({
          target: target.toSlice(),
          positional: args.positional
        });
      } else {
        return Err(generateSyntaxError("you can only yield to a literal string value", target.loc));
      }
    }
  },
  translate: function translate(_ref, _ref2) {
    var node = _ref.node,
        state = _ref.state;
    var target = _ref2.target,
        positional = _ref2.positional;
    return VISIT_EXPRS.Positional(positional, state).mapOk(function (positional) {
      return new mir.Yield({
        loc: node.loc,
        target: target,
        to: state.scope.allocateBlock(target.chars),
        positional: positional
      });
    });
  }
}).kw('debugger', {
  assert: function assert(node) {
    var args = node.args;
    var positional = args.positional;

    if (args.isEmpty()) {
      return Ok(undefined);
    } else {
      if (positional.isEmpty()) {
        return Err(generateSyntaxError("debugger does not take any named arguments", node.loc));
      } else {
        return Err(generateSyntaxError("debugger does not take any positional arguments", node.loc));
      }
    }
  },
  translate: function translate(_ref3) {
    var node = _ref3.node,
        scope = _ref3.state.scope;
    scope.setHasEval();
    return Ok(new mir.Debugger({
      loc: node.loc,
      scope: scope
    }));
  }
}).kw('component', {
  assert: assertCurryKeyword(0
  /* Component */
  ),
  translate: function translate(_ref4, _ref5) {
    var node = _ref4.node,
        state = _ref4.state;
    var definition = _ref5.definition,
        args = _ref5.args;
    var definitionResult = VISIT_EXPRS.visit(definition, state);
    var argsResult = VISIT_EXPRS.Args(args, state);
    return Result.all(definitionResult, argsResult).mapOk(function (_ref6) {
      var definition = _ref6[0],
          args = _ref6[1];
      return new mir.InvokeComponent({
        loc: node.loc,
        definition: definition,
        args: args,
        blocks: null
      });
    });
  }
}).kw('helper', {
  assert: assertCurryKeyword(1
  /* Helper */
  ),
  translate: function translate(_ref7, _ref8) {
    var node = _ref7.node,
        state = _ref7.state;
    var definition = _ref8.definition,
        args = _ref8.args;
    var definitionResult = VISIT_EXPRS.visit(definition, state);
    var argsResult = VISIT_EXPRS.Args(args, state);
    return Result.all(definitionResult, argsResult).mapOk(function (_ref9) {
      var definition = _ref9[0],
          args = _ref9[1];
      var text = new mir.CallExpression({
        callee: definition,
        args: args,
        loc: node.loc
      });
      return new mir.AppendTextNode({
        loc: node.loc,
        text: text
      });
    });
  }
});
