// @glimmer/vm/dist/modules/es2017/lib/registers.js
/**
 * Registers
 *
 * For the most part, these follows MIPS naming conventions, however the
 * register numbers are different.
 */
// $0 or $pc (program counter): pointer into `program` for the next insturction; -1 means exit
export const $pc = 0; // $1 or $ra (return address): pointer into `program` for the return

export const $ra = 1; // $2 or $fp (frame pointer): pointer into the `evalStack` for the base of the stack

export const $fp = 2; // $3 or $sp (stack pointer): pointer into the `evalStack` for the top of the stack

export const $sp = 3; // $4-$5 or $s0-$s1 (saved): callee saved general-purpose registers

export const $s0 = 4;
export const $s1 = 5; // $6-$7 or $t0-$t1 (temporaries): caller saved general-purpose registers

export const $t0 = 6;
export const $t1 = 7; // $8 or $v0 (return value)

export const $v0 = 8;
export function isLowLevelRegister(register) {
  return register <= $sp;
}
export var SavedRegister;

(function (SavedRegister) {
  SavedRegister[SavedRegister["s0"] = 4] = "s0";
  SavedRegister[SavedRegister["s1"] = 5] = "s1";
})(SavedRegister || (SavedRegister = {}));

export var TemporaryRegister;

(function (TemporaryRegister) {
  TemporaryRegister[TemporaryRegister["t0"] = 6] = "t0";
  TemporaryRegister[TemporaryRegister["t1"] = 7] = "t1";
})(TemporaryRegister || (TemporaryRegister = {}));
