// @glimmer/compiler/dist/commonjs/es5/lib/passes/2-encoding/mir.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tail = exports.Args = exports.NamedArgument = exports.NamedArguments = exports.Positional = exports.Curry = exports.HasBlockParams = exports.HasBlock = exports.InterpolateExpression = exports.Missing = exports.GetFree = exports.GetFreeWithContext = exports.GetSymbol = exports.GetWithResolver = exports.PathExpression = exports.SplatAttr = exports.InvokeBlock = exports.Modifier = exports.DeprecatedCallExpression = exports.CallExpression = exports.Debugger = exports.Yield = exports.ElementParameters = exports.SimpleElement = exports.DynamicAttr = exports.StaticAttr = exports.Component = exports.AppendComment = exports.AppendTextNode = exports.AppendTrustedHTML = exports.EndBlock = exports.NamedBlock = exports.NamedBlocks = exports.InvokeComponent = exports.Log = exports.GetDynamicVar = exports.WithDynamicVars = exports.Let = exports.With = exports.Each = exports.IfInline = exports.If = exports.Not = exports.InElement = exports.Template = void 0;

var _syntax = require("@glimmer/syntax");

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

var Template = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(Template, _node$fields);

  function Template() {
    return _node$fields.apply(this, arguments) || this;
  }

  return Template;
}((0, _syntax.node)('Template').fields());

exports.Template = Template;

var InElement = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(InElement, _node$fields2);

  function InElement() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return InElement;
}((0, _syntax.node)('InElement').fields());

exports.InElement = InElement;

var Not = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(Not, _node$fields3);

  function Not() {
    return _node$fields3.apply(this, arguments) || this;
  }

  return Not;
}((0, _syntax.node)('Not').fields());

exports.Not = Not;

var If = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(If, _node$fields4);

  function If() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return If;
}((0, _syntax.node)('If').fields());

exports.If = If;

var IfInline = /*#__PURE__*/function (_node$fields5) {
  _inheritsLoose(IfInline, _node$fields5);

  function IfInline() {
    return _node$fields5.apply(this, arguments) || this;
  }

  return IfInline;
}((0, _syntax.node)('IfInline').fields());

exports.IfInline = IfInline;

var Each = /*#__PURE__*/function (_node$fields6) {
  _inheritsLoose(Each, _node$fields6);

  function Each() {
    return _node$fields6.apply(this, arguments) || this;
  }

  return Each;
}((0, _syntax.node)('Each').fields());

exports.Each = Each;

var With = /*#__PURE__*/function (_node$fields7) {
  _inheritsLoose(With, _node$fields7);

  function With() {
    return _node$fields7.apply(this, arguments) || this;
  }

  return With;
}((0, _syntax.node)('With').fields());

exports.With = With;

var Let = /*#__PURE__*/function (_node$fields8) {
  _inheritsLoose(Let, _node$fields8);

  function Let() {
    return _node$fields8.apply(this, arguments) || this;
  }

  return Let;
}((0, _syntax.node)('Let').fields());

exports.Let = Let;

var WithDynamicVars = /*#__PURE__*/function (_node$fields9) {
  _inheritsLoose(WithDynamicVars, _node$fields9);

  function WithDynamicVars() {
    return _node$fields9.apply(this, arguments) || this;
  }

  return WithDynamicVars;
}((0, _syntax.node)('WithDynamicVars').fields());

exports.WithDynamicVars = WithDynamicVars;

var GetDynamicVar = /*#__PURE__*/function (_node$fields10) {
  _inheritsLoose(GetDynamicVar, _node$fields10);

  function GetDynamicVar() {
    return _node$fields10.apply(this, arguments) || this;
  }

  return GetDynamicVar;
}((0, _syntax.node)('GetDynamicVar').fields());

exports.GetDynamicVar = GetDynamicVar;

var Log = /*#__PURE__*/function (_node$fields11) {
  _inheritsLoose(Log, _node$fields11);

  function Log() {
    return _node$fields11.apply(this, arguments) || this;
  }

  return Log;
}((0, _syntax.node)('Log').fields());

exports.Log = Log;

var InvokeComponent = /*#__PURE__*/function (_node$fields12) {
  _inheritsLoose(InvokeComponent, _node$fields12);

  function InvokeComponent() {
    return _node$fields12.apply(this, arguments) || this;
  }

  return InvokeComponent;
}((0, _syntax.node)('InvokeComponent').fields());

exports.InvokeComponent = InvokeComponent;

var NamedBlocks = /*#__PURE__*/function (_node$fields13) {
  _inheritsLoose(NamedBlocks, _node$fields13);

  function NamedBlocks() {
    return _node$fields13.apply(this, arguments) || this;
  }

  return NamedBlocks;
}((0, _syntax.node)('NamedBlocks').fields());

exports.NamedBlocks = NamedBlocks;

var NamedBlock = /*#__PURE__*/function (_node$fields14) {
  _inheritsLoose(NamedBlock, _node$fields14);

  function NamedBlock() {
    return _node$fields14.apply(this, arguments) || this;
  }

  return NamedBlock;
}((0, _syntax.node)('NamedBlock').fields());

exports.NamedBlock = NamedBlock;

var EndBlock = /*#__PURE__*/function (_node$fields15) {
  _inheritsLoose(EndBlock, _node$fields15);

  function EndBlock() {
    return _node$fields15.apply(this, arguments) || this;
  }

  return EndBlock;
}((0, _syntax.node)('EndBlock').fields());

exports.EndBlock = EndBlock;

var AppendTrustedHTML = /*#__PURE__*/function (_node$fields16) {
  _inheritsLoose(AppendTrustedHTML, _node$fields16);

  function AppendTrustedHTML() {
    return _node$fields16.apply(this, arguments) || this;
  }

  return AppendTrustedHTML;
}((0, _syntax.node)('AppendTrustedHTML').fields());

exports.AppendTrustedHTML = AppendTrustedHTML;

var AppendTextNode = /*#__PURE__*/function (_node$fields17) {
  _inheritsLoose(AppendTextNode, _node$fields17);

  function AppendTextNode() {
    return _node$fields17.apply(this, arguments) || this;
  }

  return AppendTextNode;
}((0, _syntax.node)('AppendTextNode').fields());

exports.AppendTextNode = AppendTextNode;

var AppendComment = /*#__PURE__*/function (_node$fields18) {
  _inheritsLoose(AppendComment, _node$fields18);

  function AppendComment() {
    return _node$fields18.apply(this, arguments) || this;
  }

  return AppendComment;
}((0, _syntax.node)('AppendComment').fields());

exports.AppendComment = AppendComment;

var Component = /*#__PURE__*/function (_node$fields19) {
  _inheritsLoose(Component, _node$fields19);

  function Component() {
    return _node$fields19.apply(this, arguments) || this;
  }

  return Component;
}((0, _syntax.node)('Component').fields());

exports.Component = Component;

var StaticAttr = /*#__PURE__*/function (_node$fields20) {
  _inheritsLoose(StaticAttr, _node$fields20);

  function StaticAttr() {
    return _node$fields20.apply(this, arguments) || this;
  }

  return StaticAttr;
}((0, _syntax.node)('StaticAttr').fields());

exports.StaticAttr = StaticAttr;

var DynamicAttr = /*#__PURE__*/function (_node$fields21) {
  _inheritsLoose(DynamicAttr, _node$fields21);

  function DynamicAttr() {
    return _node$fields21.apply(this, arguments) || this;
  }

  return DynamicAttr;
}((0, _syntax.node)('DynamicAttr').fields());

exports.DynamicAttr = DynamicAttr;

var SimpleElement = /*#__PURE__*/function (_node$fields22) {
  _inheritsLoose(SimpleElement, _node$fields22);

  function SimpleElement() {
    return _node$fields22.apply(this, arguments) || this;
  }

  return SimpleElement;
}((0, _syntax.node)('SimpleElement').fields());

exports.SimpleElement = SimpleElement;

var ElementParameters = /*#__PURE__*/function (_node$fields23) {
  _inheritsLoose(ElementParameters, _node$fields23);

  function ElementParameters() {
    return _node$fields23.apply(this, arguments) || this;
  }

  return ElementParameters;
}((0, _syntax.node)('ElementParameters').fields());

exports.ElementParameters = ElementParameters;

var Yield = /*#__PURE__*/function (_node$fields24) {
  _inheritsLoose(Yield, _node$fields24);

  function Yield() {
    return _node$fields24.apply(this, arguments) || this;
  }

  return Yield;
}((0, _syntax.node)('Yield').fields());

exports.Yield = Yield;

var Debugger = /*#__PURE__*/function (_node$fields25) {
  _inheritsLoose(Debugger, _node$fields25);

  function Debugger() {
    return _node$fields25.apply(this, arguments) || this;
  }

  return Debugger;
}((0, _syntax.node)('Debugger').fields());

exports.Debugger = Debugger;

var CallExpression = /*#__PURE__*/function (_node$fields26) {
  _inheritsLoose(CallExpression, _node$fields26);

  function CallExpression() {
    return _node$fields26.apply(this, arguments) || this;
  }

  return CallExpression;
}((0, _syntax.node)('CallExpression').fields());

exports.CallExpression = CallExpression;

var DeprecatedCallExpression = /*#__PURE__*/function (_node$fields27) {
  _inheritsLoose(DeprecatedCallExpression, _node$fields27);

  function DeprecatedCallExpression() {
    return _node$fields27.apply(this, arguments) || this;
  }

  return DeprecatedCallExpression;
}((0, _syntax.node)('DeprecatedCallExpression').fields());

exports.DeprecatedCallExpression = DeprecatedCallExpression;

var Modifier = /*#__PURE__*/function (_node$fields28) {
  _inheritsLoose(Modifier, _node$fields28);

  function Modifier() {
    return _node$fields28.apply(this, arguments) || this;
  }

  return Modifier;
}((0, _syntax.node)('Modifier').fields());

exports.Modifier = Modifier;

var InvokeBlock = /*#__PURE__*/function (_node$fields29) {
  _inheritsLoose(InvokeBlock, _node$fields29);

  function InvokeBlock() {
    return _node$fields29.apply(this, arguments) || this;
  }

  return InvokeBlock;
}((0, _syntax.node)('InvokeBlock').fields());

exports.InvokeBlock = InvokeBlock;

var SplatAttr = /*#__PURE__*/function (_node$fields30) {
  _inheritsLoose(SplatAttr, _node$fields30);

  function SplatAttr() {
    return _node$fields30.apply(this, arguments) || this;
  }

  return SplatAttr;
}((0, _syntax.node)('SplatAttr').fields());

exports.SplatAttr = SplatAttr;

var PathExpression = /*#__PURE__*/function (_node$fields31) {
  _inheritsLoose(PathExpression, _node$fields31);

  function PathExpression() {
    return _node$fields31.apply(this, arguments) || this;
  }

  return PathExpression;
}((0, _syntax.node)('PathExpression').fields());

exports.PathExpression = PathExpression;

var GetWithResolver = /*#__PURE__*/function (_node$fields32) {
  _inheritsLoose(GetWithResolver, _node$fields32);

  function GetWithResolver() {
    return _node$fields32.apply(this, arguments) || this;
  }

  return GetWithResolver;
}((0, _syntax.node)('GetWithResolver').fields());

exports.GetWithResolver = GetWithResolver;

var GetSymbol = /*#__PURE__*/function (_node$fields33) {
  _inheritsLoose(GetSymbol, _node$fields33);

  function GetSymbol() {
    return _node$fields33.apply(this, arguments) || this;
  }

  return GetSymbol;
}((0, _syntax.node)('GetSymbol').fields());

exports.GetSymbol = GetSymbol;

var GetFreeWithContext = /*#__PURE__*/function (_node$fields34) {
  _inheritsLoose(GetFreeWithContext, _node$fields34);

  function GetFreeWithContext() {
    return _node$fields34.apply(this, arguments) || this;
  }

  return GetFreeWithContext;
}((0, _syntax.node)('GetFreeWithContext').fields());
/** strict mode */


exports.GetFreeWithContext = GetFreeWithContext;

var GetFree = /*#__PURE__*/function (_node$fields35) {
  _inheritsLoose(GetFree, _node$fields35);

  function GetFree() {
    return _node$fields35.apply(this, arguments) || this;
  }

  return GetFree;
}((0, _syntax.node)('GetFree').fields());

exports.GetFree = GetFree;

var Missing = /*#__PURE__*/function (_node$fields36) {
  _inheritsLoose(Missing, _node$fields36);

  function Missing() {
    return _node$fields36.apply(this, arguments) || this;
  }

  return Missing;
}((0, _syntax.node)('Missing').fields());

exports.Missing = Missing;

var InterpolateExpression = /*#__PURE__*/function (_node$fields37) {
  _inheritsLoose(InterpolateExpression, _node$fields37);

  function InterpolateExpression() {
    return _node$fields37.apply(this, arguments) || this;
  }

  return InterpolateExpression;
}((0, _syntax.node)('InterpolateExpression').fields());

exports.InterpolateExpression = InterpolateExpression;

var HasBlock = /*#__PURE__*/function (_node$fields38) {
  _inheritsLoose(HasBlock, _node$fields38);

  function HasBlock() {
    return _node$fields38.apply(this, arguments) || this;
  }

  return HasBlock;
}((0, _syntax.node)('HasBlock').fields());

exports.HasBlock = HasBlock;

var HasBlockParams = /*#__PURE__*/function (_node$fields39) {
  _inheritsLoose(HasBlockParams, _node$fields39);

  function HasBlockParams() {
    return _node$fields39.apply(this, arguments) || this;
  }

  return HasBlockParams;
}((0, _syntax.node)('HasBlockParams').fields());

exports.HasBlockParams = HasBlockParams;

var Curry = /*#__PURE__*/function (_node$fields40) {
  _inheritsLoose(Curry, _node$fields40);

  function Curry() {
    return _node$fields40.apply(this, arguments) || this;
  }

  return Curry;
}((0, _syntax.node)('Curry').fields());

exports.Curry = Curry;

var Positional = /*#__PURE__*/function (_node$fields41) {
  _inheritsLoose(Positional, _node$fields41);

  function Positional() {
    return _node$fields41.apply(this, arguments) || this;
  }

  return Positional;
}((0, _syntax.node)('Positional').fields());

exports.Positional = Positional;

var NamedArguments = /*#__PURE__*/function (_node$fields42) {
  _inheritsLoose(NamedArguments, _node$fields42);

  function NamedArguments() {
    return _node$fields42.apply(this, arguments) || this;
  }

  return NamedArguments;
}((0, _syntax.node)('NamedArguments').fields());

exports.NamedArguments = NamedArguments;

var NamedArgument = /*#__PURE__*/function (_node$fields43) {
  _inheritsLoose(NamedArgument, _node$fields43);

  function NamedArgument() {
    return _node$fields43.apply(this, arguments) || this;
  }

  return NamedArgument;
}((0, _syntax.node)('NamedArgument').fields());

exports.NamedArgument = NamedArgument;

var Args = /*#__PURE__*/function (_node$fields44) {
  _inheritsLoose(Args, _node$fields44);

  function Args() {
    return _node$fields44.apply(this, arguments) || this;
  }

  return Args;
}((0, _syntax.node)('Args').fields());

exports.Args = Args;

var Tail = /*#__PURE__*/function (_node$fields45) {
  _inheritsLoose(Tail, _node$fields45);

  function Tail() {
    return _node$fields45.apply(this, arguments) || this;
  }

  return Tail;
}((0, _syntax.node)('Tail').fields());

exports.Tail = Tail;
