// @glimmer/runtime/dist/modules/es5/lib/bounds.js
export var CursorImpl = function CursorImpl(element, nextSibling) {
  this.element = element;
  this.nextSibling = nextSibling;
};
export var ConcreteBounds = /*#__PURE__*/function () {
  function ConcreteBounds(parentNode, first, last) {
    this.parentNode = parentNode;
    this.first = first;
    this.last = last;
  }

  var _proto = ConcreteBounds.prototype;

  _proto.parentElement = function parentElement() {
    return this.parentNode;
  };

  _proto.firstNode = function firstNode() {
    return this.first;
  };

  _proto.lastNode = function lastNode() {
    return this.last;
  };

  return ConcreteBounds;
}();
export var SingleNodeBounds = /*#__PURE__*/function () {
  function SingleNodeBounds(parentNode, node) {
    this.parentNode = parentNode;
    this.node = node;
  }

  var _proto2 = SingleNodeBounds.prototype;

  _proto2.parentElement = function parentElement() {
    return this.parentNode;
  };

  _proto2.firstNode = function firstNode() {
    return this.node;
  };

  _proto2.lastNode = function lastNode() {
    return this.node;
  };

  return SingleNodeBounds;
}();
export function move(bounds, reference) {
  var parent = bounds.parentElement();
  var first = bounds.firstNode();
  var last = bounds.lastNode();
  var current = first;

  while (true) {
    var next = current.nextSibling;
    parent.insertBefore(current, reference);

    if (current === last) {
      return next;
    }

    current = next;
  }
}
export function clear(bounds) {
  var parent = bounds.parentElement();
  var first = bounds.firstNode();
  var last = bounds.lastNode();
  var current = first;

  while (true) {
    var next = current.nextSibling;
    parent.removeChild(current);

    if (current === last) {
      return next;
    }

    current = next;
  }
}
