// @glimmer/runtime/dist/modules/es5/lib/opcodes.js
import { valueForRef } from '@glimmer/reference';
import { assert, fillNulls, LOCAL_LOGGER } from '@glimmer/util';
import { $fp, $pc, $ra, $sp } from '@glimmer/vm';
import { isScopeReference } from './scope';
import { DESTROYABLE_STACK, INNER_VM, STACKS } from './symbols';
import { CURSOR_STACK } from './vm/element-builder';
export var AppendOpcodes = /*#__PURE__*/function () {
  function AppendOpcodes() {
    this.evaluateOpcode = fillNulls(104
    /* Size */
    ).slice();
  }

  var _proto = AppendOpcodes.prototype;

  _proto.add = function add(name, evaluate, kind) {
    if (kind === void 0) {
      kind = 'syscall';
    }

    this.evaluateOpcode[name] = {
      syscall: kind !== 'machine',
      evaluate: evaluate
    };
  };

  _proto.debugBefore = function debugBefore(vm, opcode) {
    var params = undefined;
    var opName = undefined;

    if (false
    /* LOCAL_SHOULD_LOG */
    ) {
        var pos = vm[INNER_VM].fetchRegister($pc) - opcode.size;
        var _ref = [];
        opName = _ref[0];
        params = _ref[1];
        // console.log(`${typePos(vm['pc'])}.`);
        LOCAL_LOGGER.log(pos + ". ");
        var debugParams = [];

        for (var prop in params) {
          debugParams.push(prop, '=', params[prop]);
        }

        LOCAL_LOGGER.log.apply(LOCAL_LOGGER, debugParams);
      }

    var sp;

    if (false
    /* LOCAL_DEBUG */
    ) {
        sp = vm.fetchValue($sp);
      }

    return {
      sp: sp,
      pc: vm.fetchValue($pc),
      name: opName,
      params: params,
      type: opcode.type,
      isMachine: opcode.isMachine,
      size: opcode.size,
      state: undefined
    };
  };

  _proto.debugAfter = function debugAfter(vm, pre) {
    var sp = pre.sp,
        type = pre.type,
        isMachine = pre.isMachine,
        pc = pre.pc;

    if (false
    /* LOCAL_DEBUG */
    ) {
        var meta = type;
        var actualChange = vm.fetchValue($sp) - sp;

        if (meta && meta.check && typeof meta.stackChange === 'number' && meta.stackChange !== actualChange) {
          throw new Error("Error in " + pre.name + ":\n\n" + pc + ". " + "\n\nStack changed by " + actualChange + ", expected " + meta.stackChange);
        }

        if (false
        /* LOCAL_SHOULD_LOG */
        ) {
            LOCAL_LOGGER.log('%c -> pc: %d, ra: %d, fp: %d, sp: %d, s0: %O, s1: %O, t0: %O, t1: %O, v0: %O', 'color: orange', vm[INNER_VM].registers[$pc], vm[INNER_VM].registers[$ra], vm[INNER_VM].registers[$fp], vm[INNER_VM].registers[$sp], vm['s0'], vm['s1'], vm['t0'], vm['t1'], vm['v0']);
            LOCAL_LOGGER.log('%c -> eval stack', 'color: red', vm.stack.toArray());
            LOCAL_LOGGER.log('%c -> block stack', 'color: magenta', vm.elements().debugBlocks());
            LOCAL_LOGGER.log('%c -> destructor stack', 'color: violet', vm[DESTROYABLE_STACK].toArray());

            if (vm[STACKS].scope.current === null) {
              LOCAL_LOGGER.log('%c -> scope', 'color: green', 'null');
            } else {
              LOCAL_LOGGER.log('%c -> scope', 'color: green', vm.scope().slots.map(function (s) {
                return isScopeReference(s) ? valueForRef(s) : s;
              }));
            }

            LOCAL_LOGGER.log('%c -> elements', 'color: blue', vm.elements()[CURSOR_STACK].current.element);
            LOCAL_LOGGER.log('%c -> constructing', 'color: aqua', vm.elements()['constructing']);
          }
      }
  };

  _proto.evaluate = function evaluate(vm, opcode, type) {
    var operation = this.evaluateOpcode[type];

    if (operation.syscall) {
      false && assert(!opcode.isMachine, "BUG: Mismatch between operation.syscall (" + operation.syscall + ") and opcode.isMachine (" + opcode.isMachine + ") for " + opcode.type);
      operation.evaluate(vm, opcode);
    } else {
      false && assert(opcode.isMachine, "BUG: Mismatch between operation.syscall (" + operation.syscall + ") and opcode.isMachine (" + opcode.isMachine + ") for " + opcode.type);
      operation.evaluate(vm[INNER_VM], opcode);
    }
  };

  return AppendOpcodes;
}();
export var APPEND_OPCODES = new AppendOpcodes();
