// @glimmer/compiler/dist/commonjs/es5/lib/passes/1-normalization/visitors/element/component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClassifiedComponent = void 0;

var mir = _interopRequireWildcard(require("../../../2-encoding/mir"));

var _expressions = require("../expressions");

var _statements = require("../statements");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var ClassifiedComponent = /*#__PURE__*/function () {
  function ClassifiedComponent(tag, element) {
    this.tag = tag;
    this.element = element;
    this.dynamicFeatures = true;
  }

  var _proto = ClassifiedComponent.prototype;

  _proto.arg = function arg(attr, _ref) {
    var state = _ref.state;
    var name = attr.name;
    return _expressions.VISIT_EXPRS.visit((0, _expressions.convertPathToCallIfKeyword)(attr.value), state).mapOk(function (value) {
      return new mir.NamedArgument({
        loc: attr.loc,
        key: name,
        value: value
      });
    });
  };

  _proto.toStatement = function toStatement(component, _ref2) {
    var _this = this;

    var args = _ref2.args,
        params = _ref2.params;
    var element = component.element,
        state = component.state;
    return this.blocks(state).mapOk(function (blocks) {
      return new mir.Component({
        loc: element.loc,
        tag: _this.tag,
        params: params,
        args: args,
        blocks: blocks
      });
    });
  };

  _proto.blocks = function blocks(state) {
    return _statements.VISIT_STMTS.NamedBlocks(this.element.blocks, state);
  };

  return ClassifiedComponent;
}();

exports.ClassifiedComponent = ClassifiedComponent;
