// @glimmer/compiler/dist/commonjs/es5/lib/passes/1-normalization/visitors/element/simple-element.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClassifiedSimpleElement = void 0;

var _syntax = require("@glimmer/syntax");

var _result = require("../../../../shared/result");

var mir = _interopRequireWildcard(require("../../../2-encoding/mir"));

var _statements = require("../statements");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var ClassifiedSimpleElement = /*#__PURE__*/function () {
  function ClassifiedSimpleElement(tag, element, dynamicFeatures) {
    this.tag = tag;
    this.element = element;
    this.dynamicFeatures = dynamicFeatures;
    this.isComponent = false;
  }

  var _proto = ClassifiedSimpleElement.prototype;

  _proto.arg = function arg(attr) {
    return (0, _result.Err)((0, _syntax.generateSyntaxError)(attr.name.chars + " is not a valid attribute name. @arguments are only allowed on components, but the tag for this element (`" + this.tag.chars + "`) is a regular, non-component HTML element.", attr.loc));
  };

  _proto.toStatement = function toStatement(classified, _ref) {
    var _this = this;

    var params = _ref.params;
    var state = classified.state,
        element = classified.element;

    var body = _statements.VISIT_STMTS.visitList(this.element.body, state);

    return body.mapOk(function (body) {
      return new mir.SimpleElement({
        loc: element.loc,
        tag: _this.tag,
        params: params,
        body: body.toArray(),
        dynamicFeatures: _this.dynamicFeatures
      });
    });
  };

  return ClassifiedSimpleElement;
}();

exports.ClassifiedSimpleElement = ClassifiedSimpleElement;
