// @glimmer/compiler/dist/commonjs/es5/lib/wire-format-debug.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("@glimmer/util");

var _utils = require("./utils");

function _createForOfIteratorHelperLoose(o, allowArrayLike) {
  var it;

  if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) {
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;
      return function () {
        if (i >= o.length) return {
          done: true
        };
        return {
          done: false,
          value: o[i++]
        };
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  it = o[Symbol.iterator]();
  return it.next.bind(it);
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var WireFormatDebugger = /*#__PURE__*/function () {
  function WireFormatDebugger(_ref) {
    var _statements = _ref[0],
        symbols = _ref[1],
        _hasEval = _ref[2],
        upvars = _ref[3];
    this.upvars = upvars;
    this.symbols = symbols;
  }

  var _proto = WireFormatDebugger.prototype;

  _proto.format = function format(program) {
    var out = [];

    for (var _iterator = _createForOfIteratorHelperLoose(program[0]), _step; !(_step = _iterator()).done;) {
      var statement = _step.value;
      out.push(this.formatOpcode(statement));
    }

    return out;
  };

  _proto.formatOpcode = function formatOpcode(opcode) {
    if (Array.isArray(opcode)) {
      switch (opcode[0]) {
        case 1
        /* Append */
        :
          return ['append', this.formatOpcode(opcode[1])];

        case 2
        /* TrustingAppend */
        :
          return ['trusting-append', this.formatOpcode(opcode[1])];

        case 6
        /* Block */
        :
          return ['block', this.formatOpcode(opcode[1]), this.formatParams(opcode[2]), this.formatHash(opcode[3]), this.formatBlocks(opcode[4])];

        case 40
        /* InElement */
        :
          return ['in-element', opcode[1], this.formatOpcode(opcode[2]), opcode[3] ? this.formatOpcode(opcode[3]) : undefined];

        case 10
        /* OpenElement */
        :
          return ['open-element', (0, _utils.inflateTagName)(opcode[1])];

        case 11
        /* OpenElementWithSplat */
        :
          return ['open-element-with-splat', (0, _utils.inflateTagName)(opcode[1])];

        case 13
        /* CloseElement */
        :
          return ['close-element'];

        case 12
        /* FlushElement */
        :
          return ['flush-element'];

        case 14
        /* StaticAttr */
        :
          return ['static-attr', (0, _utils.inflateAttrName)(opcode[1]), opcode[2], opcode[3]];

        case 24
        /* StaticComponentAttr */
        :
          return ['static-component-attr', (0, _utils.inflateAttrName)(opcode[1]), opcode[2], opcode[3]];

        case 15
        /* DynamicAttr */
        :
          return ['dynamic-attr', (0, _utils.inflateAttrName)(opcode[1]), this.formatOpcode(opcode[2]), opcode[3]];

        case 16
        /* ComponentAttr */
        :
          return ['component-attr', (0, _utils.inflateAttrName)(opcode[1]), this.formatOpcode(opcode[2]), opcode[3]];

        case 17
        /* AttrSplat */
        :
          return ['attr-splat'];

        case 18
        /* Yield */
        :
          return ['yield', opcode[1], this.formatParams(opcode[2])];

        case 20
        /* DynamicArg */
        :
          return ['dynamic-arg', opcode[1], this.formatOpcode(opcode[2])];

        case 21
        /* StaticArg */
        :
          return ['static-arg', opcode[1], this.formatOpcode(opcode[2])];

        case 22
        /* TrustingDynamicAttr */
        :
          return ['trusting-dynamic-attr', (0, _utils.inflateAttrName)(opcode[1]), this.formatOpcode(opcode[2]), opcode[3]];

        case 23
        /* TrustingComponentAttr */
        :
          return ['trusting-component-attr', (0, _utils.inflateAttrName)(opcode[1]), this.formatOpcode(opcode[2]), opcode[3]];

        case 26
        /* Debugger */
        :
          return ['debugger', opcode[1]];

        case 3
        /* Comment */
        :
          return ['comment', opcode[1]];

        case 4
        /* Modifier */
        :
          return ['modifier', this.formatOpcode(opcode[1]), this.formatParams(opcode[2]), this.formatHash(opcode[3])];

        case 8
        /* Component */
        :
          return ['component', this.formatOpcode(opcode[1]), this.formatElementParams(opcode[2]), this.formatHash(opcode[3]), this.formatBlocks(opcode[4])];

        case 48
        /* HasBlock */
        :
          return ['has-block', this.formatOpcode(opcode[1])];

        case 49
        /* HasBlockParams */
        :
          return ['has-block-params', this.formatOpcode(opcode[1])];

        case 50
        /* Curry */
        :
          return ['curry', this.formatOpcode(opcode[1]), this.formatCurryType(opcode[2]), this.formatParams(opcode[3]), this.formatHash(opcode[4])];

        case 27
        /* Undefined */
        :
          return ['undefined'];

        case 28
        /* Call */
        :
          return ['call', this.formatOpcode(opcode[1]), this.formatParams(opcode[2]), this.formatHash(opcode[3])];

        case 29
        /* Concat */
        :
          return ['concat', this.formatParams(opcode[1])];

        case 31
        /* GetStrictFree */
        :
          return ['get-strict-free', this.upvars[opcode[1]], opcode[2]];

        case 34
        /* GetFreeAsComponentOrHelperHeadOrThisFallback */
        :
          return ['GetFreeAsComponentOrHelperHeadOrThisFallback', this.upvars[opcode[1]], opcode[2]];

        case 35
        /* GetFreeAsComponentOrHelperHead */
        :
          return ['GetFreeAsComponentOrHelperHead', this.upvars[opcode[1]], opcode[2]];

        case 36
        /* GetFreeAsHelperHeadOrThisFallback */
        :
          return ['GetFreeAsHelperHeadOrThisFallback', this.upvars[opcode[1]], opcode[2]];

        case 99
        /* GetFreeAsDeprecatedHelperHeadOrThisFallback */
        :
          return ['GetFreeAsDeprecatedHelperHeadOrThisFallback', this.upvars[opcode[1]]];

        case 37
        /* GetFreeAsHelperHead */
        :
          return ['GetFreeAsHelperHead', this.upvars[opcode[1]], opcode[2]];

        case 39
        /* GetFreeAsComponentHead */
        :
          return ['GetFreeAsComponentHead', this.upvars[opcode[1]], opcode[2]];

        case 38
        /* GetFreeAsModifierHead */
        :
          return ['GetFreeAsModifierHead', this.upvars[opcode[1]], opcode[2]];

        case 30
        /* GetSymbol */
        :
          {
            if (opcode[1] === 0) {
              return ['get-symbol', 'this', opcode[2]];
            } else {
              return ['get-symbol', this.symbols[opcode[1] - 1], opcode[2]];
            }
          }

        case 32
        /* GetTemplateSymbol */
        :
          {
            return ['get-template-symbol', opcode[1], opcode[2]];
          }

        case 41
        /* If */
        :
          return ['if', this.formatOpcode(opcode[1]), this.formatBlock(opcode[2]), opcode[3] ? this.formatBlock(opcode[3]) : null];

        case 52
        /* IfInline */
        :
          return ['if-inline'];

        case 51
        /* Not */
        :
          return ['not'];

        case 42
        /* Each */
        :
          return ['each', this.formatOpcode(opcode[1]), opcode[2] ? this.formatOpcode(opcode[2]) : null, this.formatBlock(opcode[3]), opcode[4] ? this.formatBlock(opcode[4]) : null];

        case 43
        /* With */
        :
          return ['with', this.formatOpcode(opcode[1]), this.formatBlock(opcode[2]), opcode[3] ? this.formatBlock(opcode[3]) : null];

        case 44
        /* Let */
        :
          return ['let', this.formatParams(opcode[1]), this.formatBlock(opcode[2])];

        case 54
        /* Log */
        :
          return ['log', this.formatParams(opcode[1])];

        case 45
        /* WithDynamicVars */
        :
          return ['-with-dynamic-vars', this.formatHash(opcode[1]), this.formatBlock(opcode[2])];

        case 53
        /* GetDynamicVar */
        :
          return ['-get-dynamic-vars', this.formatOpcode(opcode[1])];

        case 46
        /* InvokeComponent */
        :
          return ['component', this.formatOpcode(opcode[1]), this.formatParams(opcode[2]), this.formatHash(opcode[3]), this.formatBlocks(opcode[4])];
      }
    } else {
      return opcode;
    }
  };

  _proto.formatCurryType = function formatCurryType(value) {
    switch (value) {
      case 0
      /* Component */
      :
        return 'component';

      case 1
      /* Helper */
      :
        return 'helper';

      case 2
      /* Modifier */
      :
        return 'modifier';

      default:
        throw (0, _util.exhausted)(value);
    }
  };

  _proto.formatElementParams = function formatElementParams(opcodes) {
    var _this = this;

    if (opcodes === null) return null;
    return opcodes.map(function (o) {
      return _this.formatOpcode(o);
    });
  };

  _proto.formatParams = function formatParams(opcodes) {
    var _this2 = this;

    if (opcodes === null) return null;
    return opcodes.map(function (o) {
      return _this2.formatOpcode(o);
    });
  };

  _proto.formatHash = function formatHash(hash) {
    var _this3 = this;

    if (hash === null) return null;
    return hash[0].reduce(function (accum, key, index) {
      accum[key] = _this3.formatOpcode(hash[1][index]);
      return accum;
    }, (0, _util.dict)());
  };

  _proto.formatBlocks = function formatBlocks(blocks) {
    var _this4 = this;

    if (blocks === null) return null;
    return blocks[0].reduce(function (accum, key, index) {
      accum[key] = _this4.formatBlock(blocks[1][index]);
      return accum;
    }, (0, _util.dict)());
  };

  _proto.formatBlock = function formatBlock(block) {
    var _this5 = this;

    return {
      statements: block[0].map(function (s) {
        return _this5.formatOpcode(s);
      }),
      parameters: block[1]
    };
  };

  return WireFormatDebugger;
}();

exports.default = WireFormatDebugger;
