// @glimmer/opcode-compiler/dist/modules/es5/lib/compilable-template.js
import { meta } from './opcode-builder/helpers/shared';
import { EMPTY_ARRAY } from '@glimmer/util';
import { templateCompilationContext } from './opcode-builder/context';
import { debugCompiler } from './compiler';
import { STATEMENTS } from './syntax/statements';
import { encodeOp } from './opcode-builder/encoder';
export var PLACEHOLDER_HANDLE = -1;

var CompilableTemplateImpl = /*#__PURE__*/function () {
  function CompilableTemplateImpl(statements, meta, // Part of CompilableTemplate
  symbolTable, // Used for debugging
  moduleName) {
    if (moduleName === void 0) {
      moduleName = 'plain block';
    }

    this.statements = statements;
    this.meta = meta;
    this.symbolTable = symbolTable;
    this.moduleName = moduleName;
    this.compiled = null;
  } // Part of CompilableTemplate


  var _proto = CompilableTemplateImpl.prototype;

  _proto.compile = function compile(context) {
    return maybeCompile(this, context);
  };

  return CompilableTemplateImpl;
}();

export function compilable(layout, moduleName) {
  var _layout$block = layout.block,
      statements = _layout$block[0],
      symbols = _layout$block[1],
      hasEval = _layout$block[2];
  return new CompilableTemplateImpl(statements, meta(layout), {
    symbols: symbols,
    hasEval: hasEval
  }, moduleName);
}

function maybeCompile(compilable, context) {
  if (compilable.compiled !== null) return compilable.compiled;
  compilable.compiled = PLACEHOLDER_HANDLE;
  var statements = compilable.statements,
      meta = compilable.meta;
  var result = compileStatements(statements, meta, context);
  compilable.compiled = result;
  return result;
}

export function compileStatements(statements, meta, syntaxContext) {
  var sCompiler = STATEMENTS;
  var context = templateCompilationContext(syntaxContext, meta);
  var encoder = context.encoder,
      _context$program = context.program,
      constants = _context$program.constants,
      resolver = _context$program.resolver;

  function pushOp() {
    for (var _len = arguments.length, op = new Array(_len), _key = 0; _key < _len; _key++) {
      op[_key] = arguments[_key];
    }

    encodeOp(encoder, constants, resolver, meta, op);
  }

  for (var i = 0; i < statements.length; i++) {
    sCompiler.compile(pushOp, statements[i]);
  }

  var handle = context.encoder.commit(meta.size);

  if (false
  /* LOCAL_SHOULD_LOG */
  ) {
      debugCompiler(context, handle);
    }

  return handle;
}
export function compilableBlock(block, containing) {
  return new CompilableTemplateImpl(block[0], containing, {
    parameters: block[1] || EMPTY_ARRAY
  });
}
