// @glimmer/opcode-compiler/dist/modules/es5/lib/wrapped-component.js
import { templateCompilationContext } from './opcode-builder/context';
import { meta } from './opcode-builder/helpers/shared';
import { ATTRS_BLOCK, WrappedComponent } from './opcode-builder/helpers/components';
import { debugCompiler } from './compiler';
import { encodeOp } from './opcode-builder/encoder';
export var WrappedBuilder = /*#__PURE__*/function () {
  function WrappedBuilder(layout, moduleName) {
    this.layout = layout;
    this.moduleName = moduleName;
    this.compiled = null;
    var block = layout.block;
    var symbols = block[1],
        hasEval = block[2];
    symbols = symbols.slice(); // ensure ATTRS_BLOCK is always included (only once) in the list of symbols

    var attrsBlockIndex = symbols.indexOf(ATTRS_BLOCK);

    if (attrsBlockIndex === -1) {
      this.attrsBlockNumber = symbols.push(ATTRS_BLOCK);
    } else {
      this.attrsBlockNumber = attrsBlockIndex + 1;
    }

    this.symbolTable = {
      hasEval: hasEval,
      symbols: symbols
    };
  }

  var _proto = WrappedBuilder.prototype;

  _proto.compile = function compile(syntax) {
    if (this.compiled !== null) return this.compiled;
    var m = meta(this.layout);
    var context = templateCompilationContext(syntax, m);
    var encoder = context.encoder,
        _context$program = context.program,
        constants = _context$program.constants,
        resolver = _context$program.resolver;

    function pushOp() {
      for (var _len = arguments.length, op = new Array(_len), _key = 0; _key < _len; _key++) {
        op[_key] = arguments[_key];
      }

      encodeOp(encoder, constants, resolver, m, op);
    }

    WrappedComponent(pushOp, this.layout, this.attrsBlockNumber);
    var handle = context.encoder.commit(m.size);

    if (typeof handle !== 'number') {
      return handle;
    }

    this.compiled = handle;

    if (false
    /* LOCAL_SHOULD_LOG */
    ) {
        debugCompiler(context, handle);
      }

    return handle;
  };

  return WrappedBuilder;
}();
