// @glimmer/compiler/dist/commonjs/es5/lib/passes/1-normalization/utils/is-node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasPath = hasPath;
exports.isHelperInvocation = isHelperInvocation;
exports.isSimplePath = isSimplePath;
exports.isStrictHelper = isStrictHelper;
exports.assertIsValidModifier = assertIsValidModifier;
exports.isTrustingNode = isTrustingNode;

var _syntax = require("@glimmer/syntax");

var _util = require("@glimmer/util");

function hasPath(node) {
  return node.callee.type === 'Path';
}

function isHelperInvocation(node) {
  if (!hasPath(node)) {
    return false;
  }

  return !node.args.isEmpty();
}

function isSimplePath(path) {
  if (path.type === 'Path') {
    var head = path.ref,
        parts = path.tail;
    return head.type === 'Free' && head.resolution !== _syntax.ASTv2.STRICT_RESOLUTION && parts.length === 0;
  } else {
    return false;
  }
}

function isStrictHelper(expr) {
  if (expr.callee.type !== 'Path') {
    return true;
  }

  if (expr.callee.ref.type !== 'Free') {
    return true;
  }

  return expr.callee.ref.resolution === _syntax.ASTv2.STRICT_RESOLUTION;
}

function assertIsValidModifier(helper) {
  if (isStrictHelper(helper) || isSimplePath(helper.callee)) {
    return;
  }

  throw (0, _syntax.generateSyntaxError)("`" + printPath(helper.callee) + "` is not a valid name for a modifier", helper.loc);
}

function printPath(path) {
  switch (path.type) {
    case 'Literal':
      return JSON.stringify(path.value);

    case 'Path':
      {
        var printedPath = [printPathHead(path.ref)];
        printedPath.push.apply(printedPath, path.tail.map(function (t) {
          return t.chars;
        }));
        return printedPath.join('.');
      }

    case 'Call':
      return "(" + printPath(path.callee) + " ...)";

    case 'DeprecatedCall':
      return "" + path.callee.name;

    case 'Interpolate':
      throw (0, _util.unreachable)('a concat statement cannot appear as the head of an expression');
  }
}

function printPathHead(head) {
  switch (head.type) {
    case 'Arg':
      return head.name.chars;

    case 'Free':
    case 'Local':
      return head.name;

    case 'This':
      return 'this';
  }
}
/**
 * This function is checking whether an AST node is a triple-curly, which means that it's
 * a "trusting" node. In the Handlebars AST, this is indicated by the `escaped` flag, which
 * is a bit of a double-negative, so we change the terminology here for clarity.
 */


function isTrustingNode(value) {
  if (value.type === 'AppendContent') {
    return value.trusting;
  } else {
    return false;
  }
}
