// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/visitors/element/component.js
import * as mir from '../../../2-encoding/mir';
import { convertPathToCallIfKeyword, VISIT_EXPRS } from '../expressions';
import { VISIT_STMTS } from '../statements';
export var ClassifiedComponent = /*#__PURE__*/function () {
  function ClassifiedComponent(tag, element) {
    this.tag = tag;
    this.element = element;
    this.dynamicFeatures = true;
  }

  var _proto = ClassifiedComponent.prototype;

  _proto.arg = function arg(attr, _ref) {
    var state = _ref.state;
    var name = attr.name;
    return VISIT_EXPRS.visit(convertPathToCallIfKeyword(attr.value), state).mapOk(function (value) {
      return new mir.NamedArgument({
        loc: attr.loc,
        key: name,
        value: value
      });
    });
  };

  _proto.toStatement = function toStatement(component, _ref2) {
    var _this = this;

    var args = _ref2.args,
        params = _ref2.params;
    var element = component.element,
        state = component.state;
    return this.blocks(state).mapOk(function (blocks) {
      return new mir.Component({
        loc: element.loc,
        tag: _this.tag,
        params: params,
        args: args,
        blocks: blocks
      });
    });
  };

  _proto.blocks = function blocks(state) {
    return VISIT_STMTS.NamedBlocks(this.element.blocks, state);
  };

  return ClassifiedComponent;
}();
