// @glimmer/compiler/dist/modules/es5/lib/passes/1-normalization/visitors/element/simple-element.js
import { generateSyntaxError } from '@glimmer/syntax';
import { Err } from '../../../../shared/result';
import * as mir from '../../../2-encoding/mir';
import { VISIT_STMTS } from '../statements';
export var ClassifiedSimpleElement = /*#__PURE__*/function () {
  function ClassifiedSimpleElement(tag, element, dynamicFeatures) {
    this.tag = tag;
    this.element = element;
    this.dynamicFeatures = dynamicFeatures;
    this.isComponent = false;
  }

  var _proto = ClassifiedSimpleElement.prototype;

  _proto.arg = function arg(attr) {
    return Err(generateSyntaxError(attr.name.chars + " is not a valid attribute name. @arguments are only allowed on components, but the tag for this element (`" + this.tag.chars + "`) is a regular, non-component HTML element.", attr.loc));
  };

  _proto.toStatement = function toStatement(classified, _ref) {
    var _this = this;

    var params = _ref.params;
    var state = classified.state,
        element = classified.element;
    var body = VISIT_STMTS.visitList(this.element.body, state);
    return body.mapOk(function (body) {
      return new mir.SimpleElement({
        loc: element.loc,
        tag: _this.tag,
        params: params,
        body: body.toArray(),
        dynamicFeatures: _this.dynamicFeatures
      });
    });
  };

  return ClassifiedSimpleElement;
}();
