// @glimmer/runtime/dist/commonjs/es2017/lib/vm/append.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _env = require("@glimmer/env");

var _globalContext = require("@glimmer/global-context");

var _reference = require("@glimmer/reference");

var _util = require("@glimmer/util");

var _validator = require("@glimmer/validator");

var _vm = require("@glimmer/vm");

var _destroyable = require("@glimmer/destroyable");

var _vm2 = require("../compiled/opcodes/vm");

var _opcodes = require("../opcodes");

var _scope = require("../scope");

var _symbols = require("../symbols");

var _arguments = require("./arguments");

var _lowLevel = _interopRequireDefault(require("./low-level"));

var _renderResult = _interopRequireDefault(require("./render-result"));

var _stack = _interopRequireDefault(require("./stack"));

var _update = require("./update");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _a, _b;

class Stacks {
  constructor() {
    this.scope = new _util.Stack();
    this.dynamicScope = new _util.Stack();
    this.updating = new _util.Stack();
    this.cache = new _util.Stack();
    this.list = new _util.Stack();
  }

}

class VM {
  /**
   * End of migrated.
   */
  constructor(runtime, {
    pc,
    scope,
    dynamicScope,
    stack
  }, elementStack, context) {
    this.runtime = runtime;
    this.elementStack = elementStack;
    this.context = context;
    this[_a] = new Stacks();
    this[_b] = new _util.Stack();
    this.s0 = null;
    this.s1 = null;
    this.t0 = null;
    this.t1 = null;
    this.v0 = null;
    this.resume = initVM(this.context);

    if (_env.DEBUG) {
      (0, _globalContext.assertGlobalContextWasSet)();
    }

    let evalStack = _stack.default.restore(stack);

    false && (0, _util.assert)(typeof pc === 'number', 'pc is a number');
    evalStack[_symbols.REGISTERS][_vm.$pc] = pc;
    evalStack[_symbols.REGISTERS][_vm.$sp] = stack.length - 1;
    evalStack[_symbols.REGISTERS][_vm.$fp] = -1;
    this[_symbols.HEAP] = this.program.heap;
    this[_symbols.CONSTANTS] = this.program.constants;
    this.elementStack = elementStack;

    this[_symbols.STACKS].scope.push(scope);

    this[_symbols.STACKS].dynamicScope.push(dynamicScope);

    this[_symbols.ARGS] = new _arguments.VMArgumentsImpl();
    this[_symbols.INNER_VM] = new _lowLevel.default(evalStack, this[_symbols.HEAP], runtime.program, {
      debugBefore: opcode => {
        return _opcodes.APPEND_OPCODES.debugBefore(this, opcode);
      },
      debugAfter: state => {
        _opcodes.APPEND_OPCODES.debugAfter(this, state);
      }
    }, evalStack[_symbols.REGISTERS]);
    this.destructor = {};

    this[_symbols.DESTROYABLE_STACK].push(this.destructor);
  }

  get stack() {
    return this[_symbols.INNER_VM].stack;
  }
  /* Registers */


  get pc() {
    return this[_symbols.INNER_VM].fetchRegister(_vm.$pc);
  } // Fetch a value from a register onto the stack


  fetch(register) {
    let value = this.fetchValue(register);
    this.stack.push(value);
  } // Load a value from the stack into a register


  load(register) {
    let value = this.stack.pop();
    this.loadValue(register, value);
  }

  fetchValue(register) {
    if ((0, _vm.isLowLevelRegister)(register)) {
      return this[_symbols.INNER_VM].fetchRegister(register);
    }

    switch (register) {
      case _vm.$s0:
        return this.s0;

      case _vm.$s1:
        return this.s1;

      case _vm.$t0:
        return this.t0;

      case _vm.$t1:
        return this.t1;

      case _vm.$v0:
        return this.v0;
    }
  } // Load a value into a register


  loadValue(register, value) {
    if ((0, _vm.isLowLevelRegister)(register)) {
      this[_symbols.INNER_VM].loadRegister(register, value);
    }

    switch (register) {
      case _vm.$s0:
        this.s0 = value;
        break;

      case _vm.$s1:
        this.s1 = value;
        break;

      case _vm.$t0:
        this.t0 = value;
        break;

      case _vm.$t1:
        this.t1 = value;
        break;

      case _vm.$v0:
        this.v0 = value;
        break;
    }
  }
  /**
   * Migrated to Inner
   */
  // Start a new frame and save $ra and $fp on the stack


  pushFrame() {
    this[_symbols.INNER_VM].pushFrame();
  } // Restore $ra, $sp and $fp


  popFrame() {
    this[_symbols.INNER_VM].popFrame();
  } // Jump to an address in `program`


  goto(offset) {
    this[_symbols.INNER_VM].goto(offset);
  } // Save $pc into $ra, then jump to a new address in `program` (jal in MIPS)


  call(handle) {
    this[_symbols.INNER_VM].call(handle);
  } // Put a specific `program` address in $ra


  returnTo(offset) {
    this[_symbols.INNER_VM].returnTo(offset);
  } // Return to the `program` address stored in $ra


  return() {
    this[_symbols.INNER_VM].return();
  }

  static initial(runtime, context, {
    handle,
    self,
    dynamicScope,
    treeBuilder,
    numSymbols,
    owner
  }) {
    let scope = _scope.PartialScopeImpl.root(self, numSymbols, owner);

    let state = vmState(runtime.program.heap.getaddr(handle), scope, dynamicScope);
    let vm = initVM(context)(runtime, state, treeBuilder);
    vm.pushUpdating();
    return vm;
  }

  static empty(runtime, {
    handle,
    treeBuilder,
    dynamicScope,
    owner
  }, context) {
    let vm = initVM(context)(runtime, vmState(runtime.program.heap.getaddr(handle), _scope.PartialScopeImpl.root(_reference.UNDEFINED_REFERENCE, 0, owner), dynamicScope), treeBuilder);
    vm.pushUpdating();
    return vm;
  }

  compile(block) {
    let handle = (0, _util.unwrapHandle)(block.compile(this.context));
    return handle;
  }

  get program() {
    return this.runtime.program;
  }

  get env() {
    return this.runtime.env;
  }

  captureState(args, pc = this[_symbols.INNER_VM].fetchRegister(_vm.$pc)) {
    return {
      pc,
      scope: this.scope(),
      dynamicScope: this.dynamicScope(),
      stack: this.stack.capture(args)
    };
  }

  capture(args, pc = this[_symbols.INNER_VM].fetchRegister(_vm.$pc)) {
    return new _update.ResumableVMStateImpl(this.captureState(args, pc), this.resume);
  }

  beginCacheGroup(name) {
    let opcodes = this.updating();
    let guard = new _vm2.JumpIfNotModifiedOpcode();
    opcodes.push(guard);
    opcodes.push(new _vm2.BeginTrackFrameOpcode(name));

    this[_symbols.STACKS].cache.push(guard);

    (0, _validator.beginTrackFrame)(name);
  }

  commitCacheGroup() {
    let opcodes = this.updating();

    let guard = this[_symbols.STACKS].cache.pop();

    let tag = (0, _validator.endTrackFrame)();
    opcodes.push(new _vm2.EndTrackFrameOpcode(guard));
    guard.finalize(tag, opcodes.length);
  }

  enter(args) {
    let updating = [];
    let state = this.capture(args);
    let block = this.elements().pushUpdatableBlock();
    let tryOpcode = new _update.TryOpcode(state, this.runtime, block, updating);
    this.didEnter(tryOpcode);
  }

  enterItem({
    key,
    value,
    memo
  }) {
    let {
      stack
    } = this;
    let valueRef = (0, _reference.createIteratorItemRef)(value);
    let memoRef = (0, _reference.createIteratorItemRef)(memo);
    stack.push(valueRef);
    stack.push(memoRef);
    let state = this.capture(2);
    let block = this.elements().pushUpdatableBlock();
    let opcode = new _update.ListItemOpcode(state, this.runtime, block, key, memoRef, valueRef);
    this.didEnter(opcode);
    return opcode;
  }

  registerItem(opcode) {
    this.listBlock().initializeChild(opcode);
  }

  enterList(iterableRef, offset) {
    let updating = [];

    let addr = this[_symbols.INNER_VM].target(offset);

    let state = this.capture(0, addr);
    let list = this.elements().pushBlockList(updating);
    let opcode = new _update.ListBlockOpcode(state, this.runtime, list, updating, iterableRef);

    this[_symbols.STACKS].list.push(opcode);

    this.didEnter(opcode);
  }

  didEnter(opcode) {
    this.associateDestroyable(opcode);

    this[_symbols.DESTROYABLE_STACK].push(opcode);

    this.updateWith(opcode);
    this.pushUpdating(opcode.children);
  }

  exit() {
    this[_symbols.DESTROYABLE_STACK].pop();

    this.elements().popBlock();
    this.popUpdating();
  }

  exitList() {
    this.exit();

    this[_symbols.STACKS].list.pop();
  }

  pushUpdating(list = []) {
    this[_symbols.STACKS].updating.push(list);
  }

  popUpdating() {
    return this[_symbols.STACKS].updating.pop();
  }

  updateWith(opcode) {
    this.updating().push(opcode);
  }

  listBlock() {
    return this[_symbols.STACKS].list.current;
  }

  associateDestroyable(child) {
    let parent = this[_symbols.DESTROYABLE_STACK].current;
    (0, _destroyable.associateDestroyableChild)(parent, child);
  }

  tryUpdating() {
    return this[_symbols.STACKS].updating.current;
  }

  updating() {
    return this[_symbols.STACKS].updating.current;
  }

  elements() {
    return this.elementStack;
  }

  scope() {
    return this[_symbols.STACKS].scope.current;
  }

  dynamicScope() {
    return this[_symbols.STACKS].dynamicScope.current;
  }

  pushChildScope() {
    this[_symbols.STACKS].scope.push(this.scope().child());
  }

  pushDynamicScope() {
    let child = this.dynamicScope().child();

    this[_symbols.STACKS].dynamicScope.push(child);

    return child;
  }

  pushRootScope(size, owner) {
    let scope = _scope.PartialScopeImpl.sized(size, owner);

    this[_symbols.STACKS].scope.push(scope);

    return scope;
  }

  pushScope(scope) {
    this[_symbols.STACKS].scope.push(scope);
  }

  popScope() {
    this[_symbols.STACKS].scope.pop();
  }

  popDynamicScope() {
    this[_symbols.STACKS].dynamicScope.pop();
  } /// SCOPE HELPERS


  getOwner() {
    return this.scope().owner;
  }

  getSelf() {
    return this.scope().getSelf();
  }

  referenceForSymbol(symbol) {
    return this.scope().getSymbol(symbol);
  } /// EXECUTION


  execute(initialize) {
    if (_env.DEBUG) {
      let hasErrored = true;

      try {
        let value = this._execute(initialize); // using a boolean here to avoid breaking ergonomics of "pause on uncaught exceptions"
        // which would happen with a `catch` + `throw`


        hasErrored = false;
        return value;
      } finally {
        if (hasErrored) {
          // If any existing blocks are open, due to an error or something like
          // that, we need to close them all and clean things up properly.
          let elements = this.elements();

          while (elements.hasBlocks) {
            elements.popBlock();
          } // eslint-disable-next-line no-console


          console.error(`\n\nError occurred:\n\n${(0, _validator.resetTracking)()}\n\n`);
        }
      }
    } else {
      return this._execute(initialize);
    }
  }

  _execute(initialize) {
    if (false
    /* LOCAL_SHOULD_LOG */
    ) {
        _util.LOCAL_LOGGER.log(`EXECUTING FROM ${this[_symbols.INNER_VM].fetchRegister(_vm.$pc)}`);
      }

    if (initialize) initialize(this);
    let result;

    while (true) {
      result = this.next();
      if (result.done) break;
    }

    return result.value;
  }

  next() {
    let {
      env,
      elementStack
    } = this;

    let opcode = this[_symbols.INNER_VM].nextStatement();

    let result;

    if (opcode !== null) {
      this[_symbols.INNER_VM].evaluateOuter(opcode, this);

      result = {
        done: false,
        value: null
      };
    } else {
      // Unload the stack
      this.stack.reset();
      result = {
        done: true,
        value: new _renderResult.default(env, this.popUpdating(), elementStack.popBlock(), this.destructor)
      };
    }

    return result;
  }

  bindDynamicScope(names) {
    let scope = this.dynamicScope();

    for (let i = names.length - 1; i >= 0; i--) {
      let name = names[i];
      scope.set(name, this.stack.pop());
    }
  }

}

exports.default = VM;
_a = _symbols.STACKS, _b = _symbols.DESTROYABLE_STACK;

function vmState(pc, scope, dynamicScope) {
  return {
    pc,
    scope,
    dynamicScope,
    stack: []
  };
}

function initVM(context) {
  return (runtime, state, builder) => new VM(runtime, state, builder, context);
}
