// @glimmer/runtime/dist/commonjs/es2017/lib/vm/stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vm = require("@glimmer/vm");

var _symbols = require("../symbols");

var _lowLevel = require("./low-level");

class EvaluationStackImpl {
  // fp -> sp
  constructor(stack = [], registers) {
    this.stack = stack;
    this[_symbols.REGISTERS] = registers;

    if (false
    /* LOCAL_DEBUG */
    ) {
        Object.seal(this);
      }
  }

  static restore(snapshot) {
    return new this(snapshot.slice(), (0, _lowLevel.initializeRegistersWithSP)(snapshot.length - 1));
  }

  push(value) {
    this.stack[++this[_symbols.REGISTERS][_vm.$sp]] = value;
  }

  dup(position = this[_symbols.REGISTERS][_vm.$sp]) {
    this.stack[++this[_symbols.REGISTERS][_vm.$sp]] = this.stack[position];
  }

  copy(from, to) {
    this.stack[to] = this.stack[from];
  }

  pop(n = 1) {
    let top = this.stack[this[_symbols.REGISTERS][_vm.$sp]];
    this[_symbols.REGISTERS][_vm.$sp] -= n;
    return top;
  }

  peek(offset = 0) {
    return this.stack[this[_symbols.REGISTERS][_vm.$sp] - offset];
  }

  get(offset, base = this[_symbols.REGISTERS][_vm.$fp]) {
    return this.stack[base + offset];
  }

  set(value, offset, base = this[_symbols.REGISTERS][_vm.$fp]) {
    this.stack[base + offset] = value;
  }

  slice(start, end) {
    return this.stack.slice(start, end);
  }

  capture(items) {
    let end = this[_symbols.REGISTERS][_vm.$sp] + 1;
    let start = end - items;
    return this.stack.slice(start, end);
  }

  reset() {
    this.stack.length = 0;
  }

  toArray() {
    return this.stack.slice(this[_symbols.REGISTERS][_vm.$fp], this[_symbols.REGISTERS][_vm.$sp] + 1);
  }

}

exports.default = EvaluationStackImpl;
