// @glimmer/runtime/dist/commonjs/es5/lib/curried-value.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isCurriedValue = isCurriedValue;
exports.isCurriedType = isCurriedType;
exports.resolveCurriedValue = resolveCurriedValue;
exports.curry = curry;
exports.CurriedValue = void 0;

var _util = require("@glimmer/util");

var TYPE = (0, _util.symbol)('TYPE');
var INNER = (0, _util.symbol)('INNER');
var OWNER = (0, _util.symbol)('OWNER');
var ARGS = (0, _util.symbol)('ARGS');
var RESOLVED = (0, _util.symbol)('RESOLVED');
var CURRIED_VALUES = new _util._WeakSet();

function isCurriedValue(value) {
  return CURRIED_VALUES.has(value);
}

function isCurriedType(value, type) {
  return isCurriedValue(value) && value[TYPE] === type;
}

var CurriedValue =
/** @internal */
function CurriedValue(type, inner, owner, args, resolved) {
  if (resolved === void 0) {
    resolved = false;
  }

  CURRIED_VALUES.add(this);
  this[TYPE] = type;
  this[INNER] = inner;
  this[OWNER] = owner;
  this[ARGS] = args;
  this[RESOLVED] = resolved;
};

exports.CurriedValue = CurriedValue;

function resolveCurriedValue(curriedValue) {
  var currentWrapper = curriedValue;
  var positional;
  var named;
  var definition, owner, resolved;

  while (true) {
    var _currentWrapper = currentWrapper,
        curriedArgs = _currentWrapper[ARGS],
        inner = _currentWrapper[INNER];

    if (curriedArgs !== null) {
      var curriedNamed = curriedArgs.named,
          curriedPositional = curriedArgs.positional;

      if (curriedPositional.length > 0) {
        positional = positional === undefined ? curriedPositional : curriedPositional.concat(positional);
      }

      if (named === undefined) {
        named = [];
      }

      named.unshift(curriedNamed);
    }

    if (!isCurriedValue(inner)) {
      // Save off the owner that this helper was curried with. Later on,
      // we'll fetch the value of this register and set it as the owner on the
      // new root scope.
      definition = inner;
      owner = currentWrapper[OWNER];
      resolved = currentWrapper[RESOLVED];
      break;
    }

    currentWrapper = inner;
  }

  return {
    definition: definition,
    owner: owner,
    resolved: resolved,
    positional: positional,
    named: named
  };
}

function curry(type, spec, owner, args, resolved) {
  if (resolved === void 0) {
    resolved = false;
  }

  return new CurriedValue(type, spec, owner, args, resolved);
}
