// @glimmer/runtime/dist/commonjs/es5/lib/scope.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isScopeReference = isScopeReference;
exports.PartialScopeImpl = exports.DynamicScopeImpl = void 0;

var _util = require("@glimmer/util");

var _reference = require("@glimmer/reference");

var DynamicScopeImpl = /*#__PURE__*/function () {
  function DynamicScopeImpl(bucket) {
    if (bucket) {
      this.bucket = (0, _util.assign)({}, bucket);
    } else {
      this.bucket = {};
    }
  }

  var _proto = DynamicScopeImpl.prototype;

  _proto.get = function get(key) {
    return this.bucket[key];
  };

  _proto.set = function set(key, reference) {
    return this.bucket[key] = reference;
  };

  _proto.child = function child() {
    return new DynamicScopeImpl(this.bucket);
  };

  return DynamicScopeImpl;
}();

exports.DynamicScopeImpl = DynamicScopeImpl;

function isScopeReference(s) {
  if (s === null || Array.isArray(s)) return false;
  return true;
}

var PartialScopeImpl = /*#__PURE__*/function () {
  function PartialScopeImpl( // the 0th slot is `self`
  slots, owner, callerScope, // named arguments and blocks passed to a layout that uses eval
  evalScope, // locals in scope when the partial was invoked
  partialMap) {
    this.slots = slots;
    this.owner = owner;
    this.callerScope = callerScope;
    this.evalScope = evalScope;
    this.partialMap = partialMap;
  }

  PartialScopeImpl.root = function root(self, size, owner) {
    if (size === void 0) {
      size = 0;
    }

    var refs = new Array(size + 1);

    for (var i = 0; i <= size; i++) {
      refs[i] = _reference.UNDEFINED_REFERENCE;
    }

    return new PartialScopeImpl(refs, owner, null, null, null).init({
      self: self
    });
  };

  PartialScopeImpl.sized = function sized(size, owner) {
    if (size === void 0) {
      size = 0;
    }

    var refs = new Array(size + 1);

    for (var i = 0; i <= size; i++) {
      refs[i] = _reference.UNDEFINED_REFERENCE;
    }

    return new PartialScopeImpl(refs, owner, null, null, null);
  };

  var _proto2 = PartialScopeImpl.prototype;

  _proto2.init = function init(_ref) {
    var self = _ref.self;
    this.slots[0] = self;
    return this;
  };

  _proto2.getSelf = function getSelf() {
    return this.get(0);
  };

  _proto2.getSymbol = function getSymbol(symbol) {
    return this.get(symbol);
  };

  _proto2.getBlock = function getBlock(symbol) {
    var block = this.get(symbol);
    return block === _reference.UNDEFINED_REFERENCE ? null : block;
  };

  _proto2.getEvalScope = function getEvalScope() {
    return this.evalScope;
  };

  _proto2.getPartialMap = function getPartialMap() {
    return this.partialMap;
  };

  _proto2.bind = function bind(symbol, value) {
    this.set(symbol, value);
  };

  _proto2.bindSelf = function bindSelf(self) {
    this.set(0, self);
  };

  _proto2.bindSymbol = function bindSymbol(symbol, value) {
    this.set(symbol, value);
  };

  _proto2.bindBlock = function bindBlock(symbol, value) {
    this.set(symbol, value);
  };

  _proto2.bindEvalScope = function bindEvalScope(map) {
    this.evalScope = map;
  };

  _proto2.bindPartialMap = function bindPartialMap(map) {
    this.partialMap = map;
  };

  _proto2.bindCallerScope = function bindCallerScope(scope) {
    this.callerScope = scope;
  };

  _proto2.getCallerScope = function getCallerScope() {
    return this.callerScope;
  };

  _proto2.child = function child() {
    return new PartialScopeImpl(this.slots.slice(), this.owner, this.callerScope, this.evalScope, this.partialMap);
  };

  _proto2.get = function get(index) {
    if (index >= this.slots.length) {
      throw new RangeError("BUG: cannot get $" + index + " from scope; length=" + this.slots.length);
    }

    return this.slots[index];
  };

  _proto2.set = function set(index, value) {
    if (index >= this.slots.length) {
      throw new RangeError("BUG: cannot get $" + index + " from scope; length=" + this.slots.length);
    }

    this.slots[index] = value;
  };

  return PartialScopeImpl;
}();

exports.PartialScopeImpl = PartialScopeImpl;
