// @glimmer/runtime/dist/commonjs/es5/lib/symbols.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PC = exports.ARGS = exports.CONSTANTS = exports.HEAP = exports.REGISTERS = exports.STACKS = exports.DESTROYABLE_STACK = exports.INNER_VM = void 0;

var _util = require("@glimmer/util");

// These symbols represent "friend" properties that are used inside of
// the VM in other classes, but are not intended to be a part of
// Glimmer's API.
var INNER_VM = (0, _util.symbol)('INNER_VM');
exports.INNER_VM = INNER_VM;
var DESTROYABLE_STACK = (0, _util.symbol)('DESTROYABLE_STACK');
exports.DESTROYABLE_STACK = DESTROYABLE_STACK;
var STACKS = (0, _util.symbol)('STACKS');
exports.STACKS = STACKS;
var REGISTERS = (0, _util.symbol)('REGISTERS');
exports.REGISTERS = REGISTERS;
var HEAP = (0, _util.symbol)('HEAP');
exports.HEAP = HEAP;
var CONSTANTS = (0, _util.symbol)('CONSTANTS');
exports.CONSTANTS = CONSTANTS;
var ARGS = (0, _util.symbol)('ARGS');
exports.ARGS = ARGS;
var PC = (0, _util.symbol)('PC');
exports.PC = PC;
