// @glimmer/opcode-compiler/dist/commonjs/es5/lib/compilable-template.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compilable = compilable;
exports.compileStatements = compileStatements;
exports.compilableBlock = compilableBlock;
exports.PLACEHOLDER_HANDLE = void 0;

var _shared = require("./opcode-builder/helpers/shared");

var _util = require("@glimmer/util");

var _context = require("./opcode-builder/context");

var _compiler = require("./compiler");

var _statements = require("./syntax/statements");

var _encoder = require("./opcode-builder/encoder");

var PLACEHOLDER_HANDLE = -1;
exports.PLACEHOLDER_HANDLE = PLACEHOLDER_HANDLE;

var CompilableTemplateImpl = /*#__PURE__*/function () {
  function CompilableTemplateImpl(statements, meta, // Part of CompilableTemplate
  symbolTable, // Used for debugging
  moduleName) {
    if (moduleName === void 0) {
      moduleName = 'plain block';
    }

    this.statements = statements;
    this.meta = meta;
    this.symbolTable = symbolTable;
    this.moduleName = moduleName;
    this.compiled = null;
  } // Part of CompilableTemplate


  var _proto = CompilableTemplateImpl.prototype;

  _proto.compile = function compile(context) {
    return maybeCompile(this, context);
  };

  return CompilableTemplateImpl;
}();

function compilable(layout, moduleName) {
  var _layout$block = layout.block,
      statements = _layout$block[0],
      symbols = _layout$block[1],
      hasEval = _layout$block[2];
  return new CompilableTemplateImpl(statements, (0, _shared.meta)(layout), {
    symbols: symbols,
    hasEval: hasEval
  }, moduleName);
}

function maybeCompile(compilable, context) {
  if (compilable.compiled !== null) return compilable.compiled;
  compilable.compiled = PLACEHOLDER_HANDLE;
  var statements = compilable.statements,
      meta = compilable.meta;
  var result = compileStatements(statements, meta, context);
  compilable.compiled = result;
  return result;
}

function compileStatements(statements, meta, syntaxContext) {
  var sCompiler = _statements.STATEMENTS;
  var context = (0, _context.templateCompilationContext)(syntaxContext, meta);
  var encoder = context.encoder,
      _context$program = context.program,
      constants = _context$program.constants,
      resolver = _context$program.resolver;

  function pushOp() {
    for (var _len = arguments.length, op = new Array(_len), _key = 0; _key < _len; _key++) {
      op[_key] = arguments[_key];
    }

    (0, _encoder.encodeOp)(encoder, constants, resolver, meta, op);
  }

  for (var i = 0; i < statements.length; i++) {
    sCompiler.compile(pushOp, statements[i]);
  }

  var handle = context.encoder.commit(meta.size);

  if (false
  /* LOCAL_SHOULD_LOG */
  ) {
      (0, _compiler.debugCompiler)(context, handle);
    }

  return handle;
}

function compilableBlock(block, containing) {
  return new CompilableTemplateImpl(block[0], containing, {
    parameters: block[1] || _util.EMPTY_ARRAY
  });
}
