// @glimmer/opcode-compiler/dist/commonjs/es5/lib/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namedBlocks = namedBlocks;
exports.EMPTY_BLOCKS = exports.NamedBlocksImpl = void 0;

var _util = require("@glimmer/util");

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

var NamedBlocksImpl = /*#__PURE__*/function () {
  function NamedBlocksImpl(blocks) {
    this.blocks = blocks;
    this.names = blocks ? Object.keys(blocks) : [];
  }

  var _proto = NamedBlocksImpl.prototype;

  _proto.get = function get(name) {
    if (!this.blocks) return null;
    return this.blocks[name] || null;
  };

  _proto.has = function has(name) {
    var blocks = this.blocks;
    return blocks !== null && name in blocks;
  };

  _proto["with"] = function _with(name, block) {
    var blocks = this.blocks;

    if (blocks) {
      var _assign;

      return new NamedBlocksImpl((0, _util.assign)({}, blocks, (_assign = {}, _assign[name] = block, _assign)));
    } else {
      var _NamedBlocksImpl;

      return new NamedBlocksImpl((_NamedBlocksImpl = {}, _NamedBlocksImpl[name] = block, _NamedBlocksImpl));
    }
  };

  _createClass(NamedBlocksImpl, [{
    key: "hasAny",
    get: function get() {
      return this.blocks !== null;
    }
  }]);

  return NamedBlocksImpl;
}();

exports.NamedBlocksImpl = NamedBlocksImpl;
var EMPTY_BLOCKS = new NamedBlocksImpl(null);
exports.EMPTY_BLOCKS = EMPTY_BLOCKS;

function namedBlocks(blocks) {
  if (blocks === null) {
    return EMPTY_BLOCKS;
  }

  var out = (0, _util.dict)();
  var keys = blocks[0],
      values = blocks[1];

  for (var i = 0; i < keys.length; i++) {
    out[keys[i]] = values[i];
  }

  return new NamedBlocksImpl(out);
}
