// @glimmer/opcode-compiler/dist/commonjs/es5/lib/wrapped-component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WrappedBuilder = void 0;

var _context = require("./opcode-builder/context");

var _shared = require("./opcode-builder/helpers/shared");

var _components = require("./opcode-builder/helpers/components");

var _compiler = require("./compiler");

var _encoder = require("./opcode-builder/encoder");

var WrappedBuilder = /*#__PURE__*/function () {
  function WrappedBuilder(layout, moduleName) {
    this.layout = layout;
    this.moduleName = moduleName;
    this.compiled = null;
    var block = layout.block;
    var symbols = block[1],
        hasEval = block[2];
    symbols = symbols.slice(); // ensure ATTRS_BLOCK is always included (only once) in the list of symbols

    var attrsBlockIndex = symbols.indexOf(_components.ATTRS_BLOCK);

    if (attrsBlockIndex === -1) {
      this.attrsBlockNumber = symbols.push(_components.ATTRS_BLOCK);
    } else {
      this.attrsBlockNumber = attrsBlockIndex + 1;
    }

    this.symbolTable = {
      hasEval: hasEval,
      symbols: symbols
    };
  }

  var _proto = WrappedBuilder.prototype;

  _proto.compile = function compile(syntax) {
    if (this.compiled !== null) return this.compiled;
    var m = (0, _shared.meta)(this.layout);
    var context = (0, _context.templateCompilationContext)(syntax, m);
    var encoder = context.encoder,
        _context$program = context.program,
        constants = _context$program.constants,
        resolver = _context$program.resolver;

    function pushOp() {
      for (var _len = arguments.length, op = new Array(_len), _key = 0; _key < _len; _key++) {
        op[_key] = arguments[_key];
      }

      (0, _encoder.encodeOp)(encoder, constants, resolver, m, op);
    }

    (0, _components.WrappedComponent)(pushOp, this.layout, this.attrsBlockNumber);
    var handle = context.encoder.commit(m.size);

    if (typeof handle !== 'number') {
      return handle;
    }

    this.compiled = handle;

    if (false
    /* LOCAL_SHOULD_LOG */
    ) {
        (0, _compiler.debugCompiler)(context, handle);
      }

    return handle;
  };

  return WrappedBuilder;
}();

exports.WrappedBuilder = WrappedBuilder;
