// @glimmer/runtime/dist/commonjs/es2017/lib/curried-value.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isCurriedValue = isCurriedValue;
exports.isCurriedType = isCurriedType;
exports.resolveCurriedValue = resolveCurriedValue;
exports.curry = curry;
exports.CurriedValue = void 0;

var _util = require("@glimmer/util");

const TYPE = (0, _util.symbol)('TYPE');
const INNER = (0, _util.symbol)('INNER');
const OWNER = (0, _util.symbol)('OWNER');
const ARGS = (0, _util.symbol)('ARGS');
const RESOLVED = (0, _util.symbol)('RESOLVED');
const CURRIED_VALUES = new _util._WeakSet();

function isCurriedValue(value) {
  return CURRIED_VALUES.has(value);
}

function isCurriedType(value, type) {
  return isCurriedValue(value) && value[TYPE] === type;
}

class CurriedValue {
  /** @internal */
  constructor(type, inner, owner, args, resolved = false) {
    CURRIED_VALUES.add(this);
    this[TYPE] = type;
    this[INNER] = inner;
    this[OWNER] = owner;
    this[ARGS] = args;
    this[RESOLVED] = resolved;
  }

}

exports.CurriedValue = CurriedValue;

function resolveCurriedValue(curriedValue) {
  let currentWrapper = curriedValue;
  let positional;
  let named;
  let definition, owner, resolved;

  while (true) {
    let {
      [ARGS]: curriedArgs,
      [INNER]: inner
    } = currentWrapper;

    if (curriedArgs !== null) {
      let {
        named: curriedNamed,
        positional: curriedPositional
      } = curriedArgs;

      if (curriedPositional.length > 0) {
        positional = positional === undefined ? curriedPositional : curriedPositional.concat(positional);
      }

      if (named === undefined) {
        named = [];
      }

      named.unshift(curriedNamed);
    }

    if (!isCurriedValue(inner)) {
      // Save off the owner that this helper was curried with. Later on,
      // we'll fetch the value of this register and set it as the owner on the
      // new root scope.
      definition = inner;
      owner = currentWrapper[OWNER];
      resolved = currentWrapper[RESOLVED];
      break;
    }

    currentWrapper = inner;
  }

  return {
    definition,
    owner,
    resolved,
    positional,
    named
  };
}

function curry(type, spec, owner, args, resolved = false) {
  return new CurriedValue(type, spec, owner, args, resolved);
}
