// @glimmer/runtime/dist/commonjs/es2017/lib/scope.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isScopeReference = isScopeReference;
exports.PartialScopeImpl = exports.DynamicScopeImpl = void 0;

var _util = require("@glimmer/util");

var _reference = require("@glimmer/reference");

class DynamicScopeImpl {
  constructor(bucket) {
    if (bucket) {
      this.bucket = (0, _util.assign)({}, bucket);
    } else {
      this.bucket = {};
    }
  }

  get(key) {
    return this.bucket[key];
  }

  set(key, reference) {
    return this.bucket[key] = reference;
  }

  child() {
    return new DynamicScopeImpl(this.bucket);
  }

}

exports.DynamicScopeImpl = DynamicScopeImpl;

function isScopeReference(s) {
  if (s === null || Array.isArray(s)) return false;
  return true;
}

class PartialScopeImpl {
  constructor( // the 0th slot is `self`
  slots, owner, callerScope, // named arguments and blocks passed to a layout that uses eval
  evalScope, // locals in scope when the partial was invoked
  partialMap) {
    this.slots = slots;
    this.owner = owner;
    this.callerScope = callerScope;
    this.evalScope = evalScope;
    this.partialMap = partialMap;
  }

  static root(self, size = 0, owner) {
    let refs = new Array(size + 1);

    for (let i = 0; i <= size; i++) {
      refs[i] = _reference.UNDEFINED_REFERENCE;
    }

    return new PartialScopeImpl(refs, owner, null, null, null).init({
      self
    });
  }

  static sized(size = 0, owner) {
    let refs = new Array(size + 1);

    for (let i = 0; i <= size; i++) {
      refs[i] = _reference.UNDEFINED_REFERENCE;
    }

    return new PartialScopeImpl(refs, owner, null, null, null);
  }

  init({
    self
  }) {
    this.slots[0] = self;
    return this;
  }

  getSelf() {
    return this.get(0);
  }

  getSymbol(symbol) {
    return this.get(symbol);
  }

  getBlock(symbol) {
    let block = this.get(symbol);
    return block === _reference.UNDEFINED_REFERENCE ? null : block;
  }

  getEvalScope() {
    return this.evalScope;
  }

  getPartialMap() {
    return this.partialMap;
  }

  bind(symbol, value) {
    this.set(symbol, value);
  }

  bindSelf(self) {
    this.set(0, self);
  }

  bindSymbol(symbol, value) {
    this.set(symbol, value);
  }

  bindBlock(symbol, value) {
    this.set(symbol, value);
  }

  bindEvalScope(map) {
    this.evalScope = map;
  }

  bindPartialMap(map) {
    this.partialMap = map;
  }

  bindCallerScope(scope) {
    this.callerScope = scope;
  }

  getCallerScope() {
    return this.callerScope;
  }

  child() {
    return new PartialScopeImpl(this.slots.slice(), this.owner, this.callerScope, this.evalScope, this.partialMap);
  }

  get(index) {
    if (index >= this.slots.length) {
      throw new RangeError(`BUG: cannot get $${index} from scope; length=${this.slots.length}`);
    }

    return this.slots[index];
  }

  set(index, value) {
    if (index >= this.slots.length) {
      throw new RangeError(`BUG: cannot get $${index} from scope; length=${this.slots.length}`);
    }

    this.slots[index] = value;
  }

}

exports.PartialScopeImpl = PartialScopeImpl;
