// @glimmer/compiler/dist/commonjs/es5/lib/passes/1-normalization/keywords/utils/curry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertCurryKeyword = assertCurryKeyword;
exports.curryKeyword = curryKeyword;

var _syntax = require("@glimmer/syntax");

var _result = require("../../../../shared/result");

var mir = _interopRequireWildcard(require("../../../2-encoding/mir"));

var _expressions = require("../../visitors/expressions");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _CurriedTypeToReadabl;

var CurriedTypeToReadableType = (_CurriedTypeToReadabl = {}, _CurriedTypeToReadabl[0
/* Component */
] = 'component', _CurriedTypeToReadabl[1
/* Helper */
] = 'helper', _CurriedTypeToReadabl[2
/* Modifier */
] = 'modifier', _CurriedTypeToReadabl);

function assertCurryKeyword(curriedType) {
  return function (node, state) {
    var readableType = CurriedTypeToReadableType[curriedType];
    var stringsAllowed = curriedType === 0
    /* Component */
    ;
    var args = node.args;
    var definition = args.nth(0);

    if (definition === null) {
      return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + readableType + ") requires a " + readableType + " definition or identifier as its first positional parameter, did not receive any parameters.", args.loc));
    }

    if (definition.type === 'Literal') {
      if (stringsAllowed && state.isStrict) {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + readableType + ") cannot resolve string values in strict mode templates", node.loc));
      } else if (!stringsAllowed) {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + readableType + ") cannot resolve string values, you must pass a " + readableType + " definition directly", node.loc));
      }
    }

    args = new _syntax.ASTv2.Args({
      positional: new _syntax.ASTv2.PositionalArguments({
        exprs: args.positional.exprs.slice(1),
        loc: args.positional.loc
      }),
      named: args.named,
      loc: args.loc
    });
    return (0, _result.Ok)({
      definition: definition,
      args: args
    });
  };
}

function translateCurryKeyword(curriedType) {
  return function (_ref, _ref2) {
    var node = _ref.node,
        state = _ref.state;
    var definition = _ref2.definition,
        args = _ref2.args;

    var definitionResult = _expressions.VISIT_EXPRS.visit(definition, state);

    var argsResult = _expressions.VISIT_EXPRS.Args(args, state);

    return _result.Result.all(definitionResult, argsResult).mapOk(function (_ref3) {
      var definition = _ref3[0],
          args = _ref3[1];
      return new mir.Curry({
        loc: node.loc,
        curriedType: curriedType,
        definition: definition,
        args: args
      });
    });
  };
}

function curryKeyword(curriedType) {
  return {
    assert: assertCurryKeyword(curriedType),
    translate: translateCurryKeyword(curriedType)
  };
}
