// @glimmer/compiler/dist/commonjs/es5/lib/passes/1-normalization/keywords/utils/has-block.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasBlockKeyword = hasBlockKeyword;

var _syntax = require("@glimmer/syntax");

var _result = require("../../../../shared/result");

var mir = _interopRequireWildcard(require("../../../2-encoding/mir"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function assertHasBlockKeyword(type) {
  return function (node) {
    var call = node.type === 'AppendContent' ? node.value : node;
    var named = call.type === 'Call' ? call.args.named : null;
    var positionals = call.type === 'Call' ? call.args.positional : null;

    if (named && !named.isEmpty()) {
      return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + type + ") does not take any named arguments", call.loc));
    }

    if (!positionals || positionals.isEmpty()) {
      return (0, _result.Ok)(_syntax.SourceSlice.synthetic('default'));
    } else if (positionals.exprs.length === 1) {
      var positional = positionals.exprs[0];

      if (_syntax.ASTv2.isLiteral(positional, 'string')) {
        return (0, _result.Ok)(positional.toSlice());
      } else {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + type + ") can only receive a string literal as its first argument", call.loc));
      }
    } else {
      return (0, _result.Err)((0, _syntax.generateSyntaxError)("(" + type + ") only takes a single positional argument", call.loc));
    }
  };
}

function translateHasBlockKeyword(type) {
  return function (_ref, target) {
    var node = _ref.node,
        scope = _ref.state.scope;
    var block = type === 'has-block' ? new mir.HasBlock({
      loc: node.loc,
      target: target,
      symbol: scope.allocateBlock(target.chars)
    }) : new mir.HasBlockParams({
      loc: node.loc,
      target: target,
      symbol: scope.allocateBlock(target.chars)
    });
    return (0, _result.Ok)(block);
  };
}

function hasBlockKeyword(type) {
  return {
    assert: assertHasBlockKeyword(type),
    translate: translateHasBlockKeyword(type)
  };
}
