// @glimmer/compiler/dist/modules/es2017/lib/shared/list.js
import { isPresent, mapPresent } from '@glimmer/util';
export class PresentList {
  constructor(list) {
    this.list = list;
  }

  toArray() {
    return this.list;
  }

  map(callback) {
    let result = mapPresent(this.list, callback);
    return new PresentList(result);
  }

  filter(predicate) {
    let out = [];

    for (let item of this.list) {
      if (predicate(item)) {
        out.push(item);
      }
    }

    return OptionalList(out);
  }

  toPresentArray() {
    return this.list;
  }

  into({
    ifPresent
  }) {
    return ifPresent(this);
  }

}
export class EmptyList {
  constructor() {
    this.list = [];
  }

  map(_callback) {
    return new EmptyList();
  }

  filter(_predicate) {
    return new EmptyList();
  }

  toArray() {
    return this.list;
  }

  toPresentArray() {
    return null;
  }

  into({
    ifEmpty
  }) {
    return ifEmpty();
  }

} // export type OptionalList<T> = PresentList<T> | EmptyList<T>;

export function OptionalList(value) {
  if (isPresent(value)) {
    return new PresentList(value);
  } else {
    return new EmptyList();
  }
}
