// @glimmer/runtime/dist/commonjs/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "clear", {
  enumerable: true,
  get: function () {
    return _bounds.clear;
  }
});
Object.defineProperty(exports, "ConcreteBounds", {
  enumerable: true,
  get: function () {
    return _bounds.ConcreteBounds;
  }
});
Object.defineProperty(exports, "CursorImpl", {
  enumerable: true,
  get: function () {
    return _bounds.CursorImpl;
  }
});
Object.defineProperty(exports, "resetDebuggerCallback", {
  enumerable: true,
  get: function () {
    return _debugger.resetDebuggerCallback;
  }
});
Object.defineProperty(exports, "setDebuggerCallback", {
  enumerable: true,
  get: function () {
    return _debugger.setDebuggerCallback;
  }
});
Object.defineProperty(exports, "curry", {
  enumerable: true,
  get: function () {
    return _curriedValue.curry;
  }
});
Object.defineProperty(exports, "CurriedValue", {
  enumerable: true,
  get: function () {
    return _curriedValue.CurriedValue;
  }
});
Object.defineProperty(exports, "TemplateOnlyComponentManager", {
  enumerable: true,
  get: function () {
    return _templateOnly.TemplateOnlyComponentManager;
  }
});
Object.defineProperty(exports, "TEMPLATE_ONLY_COMPONENT_MANAGER", {
  enumerable: true,
  get: function () {
    return _templateOnly.TEMPLATE_ONLY_COMPONENT_MANAGER;
  }
});
Object.defineProperty(exports, "TemplateOnlyComponent", {
  enumerable: true,
  get: function () {
    return _templateOnly.TemplateOnlyComponentDefinition;
  }
});
Object.defineProperty(exports, "templateOnlyComponent", {
  enumerable: true,
  get: function () {
    return _templateOnly.templateOnlyComponent;
  }
});
Object.defineProperty(exports, "DOMChanges", {
  enumerable: true,
  get: function () {
    return _helper.default;
  }
});
Object.defineProperty(exports, "IDOMChanges", {
  enumerable: true,
  get: function () {
    return _helper.DOMChangesImpl;
  }
});
Object.defineProperty(exports, "DOMTreeConstruction", {
  enumerable: true,
  get: function () {
    return _helper.DOMTreeConstruction;
  }
});
Object.defineProperty(exports, "isWhitespace", {
  enumerable: true,
  get: function () {
    return _helper.isWhitespace;
  }
});
Object.defineProperty(exports, "normalizeProperty", {
  enumerable: true,
  get: function () {
    return _props.normalizeProperty;
  }
});
Object.defineProperty(exports, "DynamicScopeImpl", {
  enumerable: true,
  get: function () {
    return _scope.DynamicScopeImpl;
  }
});
Object.defineProperty(exports, "PartialScopeImpl", {
  enumerable: true,
  get: function () {
    return _scope.PartialScopeImpl;
  }
});
Object.defineProperty(exports, "runtimeContext", {
  enumerable: true,
  get: function () {
    return _environment.runtimeContext;
  }
});
Object.defineProperty(exports, "EnvironmentImpl", {
  enumerable: true,
  get: function () {
    return _environment.EnvironmentImpl;
  }
});
Object.defineProperty(exports, "inTransaction", {
  enumerable: true,
  get: function () {
    return _environment.inTransaction;
  }
});
Object.defineProperty(exports, "renderComponent", {
  enumerable: true,
  get: function () {
    return _render.renderComponent;
  }
});
Object.defineProperty(exports, "renderMain", {
  enumerable: true,
  get: function () {
    return _render.renderMain;
  }
});
Object.defineProperty(exports, "renderSync", {
  enumerable: true,
  get: function () {
    return _render.renderSync;
  }
});
Object.defineProperty(exports, "UpdatingVM", {
  enumerable: true,
  get: function () {
    return _vm.UpdatingVM;
  }
});
Object.defineProperty(exports, "LowLevelVM", {
  enumerable: true,
  get: function () {
    return _vm.VM;
  }
});
Object.defineProperty(exports, "EMPTY_ARGS", {
  enumerable: true,
  get: function () {
    return _arguments.EMPTY_ARGS;
  }
});
Object.defineProperty(exports, "EMPTY_NAMED", {
  enumerable: true,
  get: function () {
    return _arguments.EMPTY_NAMED;
  }
});
Object.defineProperty(exports, "EMPTY_POSITIONAL", {
  enumerable: true,
  get: function () {
    return _arguments.EMPTY_POSITIONAL;
  }
});
Object.defineProperty(exports, "createCapturedArgs", {
  enumerable: true,
  get: function () {
    return _arguments.createCapturedArgs;
  }
});
Object.defineProperty(exports, "reifyArgs", {
  enumerable: true,
  get: function () {
    return _arguments.reifyArgs;
  }
});
Object.defineProperty(exports, "reifyNamed", {
  enumerable: true,
  get: function () {
    return _arguments.reifyNamed;
  }
});
Object.defineProperty(exports, "reifyPositional", {
  enumerable: true,
  get: function () {
    return _arguments.reifyPositional;
  }
});
Object.defineProperty(exports, "DynamicAttribute", {
  enumerable: true,
  get: function () {
    return _dynamic.DynamicAttribute;
  }
});
Object.defineProperty(exports, "dynamicAttribute", {
  enumerable: true,
  get: function () {
    return _dynamic.dynamicAttribute;
  }
});
Object.defineProperty(exports, "SimpleDynamicAttribute", {
  enumerable: true,
  get: function () {
    return _dynamic.SimpleDynamicAttribute;
  }
});
Object.defineProperty(exports, "clientBuilder", {
  enumerable: true,
  get: function () {
    return _elementBuilder.clientBuilder;
  }
});
Object.defineProperty(exports, "NewElementBuilder", {
  enumerable: true,
  get: function () {
    return _elementBuilder.NewElementBuilder;
  }
});
Object.defineProperty(exports, "UpdatableBlockImpl", {
  enumerable: true,
  get: function () {
    return _elementBuilder.UpdatableBlockImpl;
  }
});
Object.defineProperty(exports, "RemoteLiveBlock", {
  enumerable: true,
  get: function () {
    return _elementBuilder.RemoteLiveBlock;
  }
});
Object.defineProperty(exports, "isSerializationFirstNode", {
  enumerable: true,
  get: function () {
    return _rehydrateBuilder.isSerializationFirstNode;
  }
});
Object.defineProperty(exports, "RehydrateBuilder", {
  enumerable: true,
  get: function () {
    return _rehydrateBuilder.RehydrateBuilder;
  }
});
Object.defineProperty(exports, "rehydrationBuilder", {
  enumerable: true,
  get: function () {
    return _rehydrateBuilder.rehydrationBuilder;
  }
});
Object.defineProperty(exports, "SERIALIZATION_FIRST_NODE_STRING", {
  enumerable: true,
  get: function () {
    return _rehydrateBuilder.SERIALIZATION_FIRST_NODE_STRING;
  }
});
Object.defineProperty(exports, "invokeHelper", {
  enumerable: true,
  get: function () {
    return _invoke.invokeHelper;
  }
});
Object.defineProperty(exports, "fn", {
  enumerable: true,
  get: function () {
    return _fn.default;
  }
});
Object.defineProperty(exports, "hash", {
  enumerable: true,
  get: function () {
    return _hash.default;
  }
});
Object.defineProperty(exports, "array", {
  enumerable: true,
  get: function () {
    return _array.default;
  }
});
Object.defineProperty(exports, "get", {
  enumerable: true,
  get: function () {
    return _get.default;
  }
});
Object.defineProperty(exports, "concat", {
  enumerable: true,
  get: function () {
    return _concat.default;
  }
});
Object.defineProperty(exports, "on", {
  enumerable: true,
  get: function () {
    return _on.default;
  }
});
Object.defineProperty(exports, "destroy", {
  enumerable: true,
  get: function () {
    return _destroyable.destroy;
  }
});
Object.defineProperty(exports, "registerDestructor", {
  enumerable: true,
  get: function () {
    return _destroyable.registerDestructor;
  }
});
Object.defineProperty(exports, "isDestroying", {
  enumerable: true,
  get: function () {
    return _destroyable.isDestroying;
  }
});
Object.defineProperty(exports, "isDestroyed", {
  enumerable: true,
  get: function () {
    return _destroyable.isDestroyed;
  }
});

require("./lib/bootstrap");

var _bounds = require("./lib/bounds");

var _debugger = require("./lib/compiled/opcodes/debugger");

var _curriedValue = require("./lib/curried-value");

var _templateOnly = require("./lib/component/template-only");

var _helper = _interopRequireWildcard(require("./lib/dom/helper"));

var _props = require("./lib/dom/props");

var _scope = require("./lib/scope");

var _environment = require("./lib/environment");

var _render = require("./lib/render");

var _vm = require("./lib/vm");

var _arguments = require("./lib/vm/arguments");

var _dynamic = require("./lib/vm/attributes/dynamic");

var _elementBuilder = require("./lib/vm/element-builder");

var _rehydrateBuilder = require("./lib/vm/rehydrate-builder");

var _invoke = require("./lib/helpers/invoke");

var _fn = _interopRequireDefault(require("./lib/helpers/fn"));

var _hash = _interopRequireDefault(require("./lib/helpers/hash"));

var _array = _interopRequireDefault(require("./lib/helpers/array"));

var _get = _interopRequireDefault(require("./lib/helpers/get"));

var _concat = _interopRequireDefault(require("./lib/helpers/concat"));

var _on = _interopRequireDefault(require("./lib/modifiers/on"));

var _destroyable = require("@glimmer/destroyable");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
