// @glimmer/wire-format/dist/commonjs/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.is = is;
exports.isAttribute = isAttribute;
exports.isStringLiteral = isStringLiteral;
exports.getStringFromValue = getStringFromValue;
exports.isArgument = isArgument;
exports.isHelper = isHelper;
exports.isGet = exports.isFlushElement = void 0;

function is(variant) {
  return function (value) {
    return Array.isArray(value) && value[0] === variant;
  };
} // Statements


var isFlushElement = is(12
/* FlushElement */
);
exports.isFlushElement = isFlushElement;

function isAttribute(val) {
  return val[0] === 14
  /* StaticAttr */
  || val[0] === 15
  /* DynamicAttr */
  || val[0] === 22
  /* TrustingDynamicAttr */
  || val[0] === 16
  /* ComponentAttr */
  || val[0] === 24
  /* StaticComponentAttr */
  || val[0] === 23
  /* TrustingComponentAttr */
  || val[0] === 17
  /* AttrSplat */
  || val[0] === 4
  /* Modifier */
  ;
}

function isStringLiteral(expr) {
  return typeof expr === 'string';
}

function getStringFromValue(expr) {
  return expr;
}

function isArgument(val) {
  return val[0] === 21
  /* StaticArg */
  || val[0] === 20
  /* DynamicArg */
  ;
}

function isHelper(expr) {
  return Array.isArray(expr) && expr[0] === 28
  /* Call */
  ;
} // Expressions


var isGet = is(30
/* GetSymbol */
);
exports.isGet = isGet;
