// @glimmer/opcode-compiler/dist/commonjs/es5/lib/opcode-builder/helpers/blocks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.YieldBlock = YieldBlock;
exports.PushYieldableBlock = PushYieldableBlock;
exports.InvokeStaticBlock = InvokeStaticBlock;
exports.InvokeStaticBlockWithStack = InvokeStaticBlockWithStack;
exports.PushSymbolTable = PushSymbolTable;
exports.PushCompilable = PushCompilable;

var _vm = require("@glimmer/vm");

var _vm2 = require("./vm");

var _operands = require("../operands");

var _shared = require("./shared");

/**
 * Yield to a block located at a particular symbol location.
 *
 * @param to the symbol containing the block to yield to
 * @param params optional block parameters to yield to the block
 */
function YieldBlock(op, to, positional) {
  (0, _shared.SimpleArgs)(op, positional, null, true);
  op(23
  /* GetBlock */
  , to);
  op(24
  /* SpreadBlock */
  );
  op(61
  /* CompileBlock */
  );
  op(64
  /* InvokeYield */
  );
  op(40
  /* PopScope */
  );
  op(1
  /* PopFrame */
  );
}
/**
 * Push an (optional) yieldable block onto the stack. The yieldable block must be known
 * statically at compile time.
 *
 * @param block An optional Compilable block
 */


function PushYieldableBlock(op, block) {
  PushSymbolTable(op, block && block[1]);
  op(62
  /* PushBlockScope */
  );
  PushCompilable(op, block);
}
/**
 * Invoke a block that is known statically at compile time.
 *
 * @param block a Compilable block
 */


function InvokeStaticBlock(op, block) {
  op(0
  /* PushFrame */
  );
  PushCompilable(op, block);
  op(61
  /* CompileBlock */
  );
  op(2
  /* InvokeVirtual */
  );
  op(1
  /* PopFrame */
  );
}
/**
 * Invoke a static block, preserving some number of stack entries for use in
 * updating.
 *
 * @param block A compilable block
 * @param callerCount A number of stack entries to preserve
 */


function InvokeStaticBlockWithStack(op, block, callerCount) {
  var parameters = block[1];
  var calleeCount = parameters.length;
  var count = Math.min(callerCount, calleeCount);

  if (count === 0) {
    InvokeStaticBlock(op, block);
    return;
  }

  op(0
  /* PushFrame */
  );

  if (count) {
    op(39
    /* ChildScope */
    );

    for (var i = 0; i < count; i++) {
      op(33
      /* Dup */
      , _vm.$fp, callerCount - i);
      op(19
      /* SetVariable */
      , parameters[i]);
    }
  }

  PushCompilable(op, block);
  op(61
  /* CompileBlock */
  );
  op(2
  /* InvokeVirtual */
  );

  if (count) {
    op(40
    /* PopScope */
    );
  }

  op(1
  /* PopFrame */
  );
}

function PushSymbolTable(op, parameters) {
  if (parameters !== null) {
    op(63
    /* PushSymbolTable */
    , (0, _operands.symbolTableOperand)({
      parameters: parameters
    }));
  } else {
    (0, _vm2.PushPrimitive)(op, null);
  }
}

function PushCompilable(op, _block) {
  if (_block === null) {
    (0, _vm2.PushPrimitive)(op, null);
  } else {
    op(28
    /* Constant */
    , (0, _operands.blockOperand)(_block));
  }
}
