// @glimmer/opcode-compiler/dist/commonjs/es5/lib/opcode-builder/helpers/shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompileArgs = CompileArgs;
exports.SimpleArgs = SimpleArgs;
exports.CompilePositional = CompilePositional;
exports.meta = meta;
exports.evalSymbols = evalSymbols;

var _util = require("@glimmer/util");

var _blocks = require("./blocks");

var _expr = require("./expr");

/**
 * Compile arguments, pushing an Arguments object onto the stack.
 *
 * @param args.params
 * @param args.hash
 * @param args.blocks
 * @param args.atNames
 */
function CompileArgs(op, positional, named, blocks, atNames) {
  var blockNames = blocks.names;

  for (var i = 0; i < blockNames.length; i++) {
    (0, _blocks.PushYieldableBlock)(op, blocks.get(blockNames[i]));
  }

  var count = CompilePositional(op, positional);
  var flags = count << 4;
  if (atNames) flags |= 8;

  if (blocks) {
    flags |= 7;
  }

  var names = _util.EMPTY_ARRAY;

  if (named) {
    names = named[0];
    var val = named[1];

    for (var _i = 0; _i < val.length; _i++) {
      (0, _expr.expr)(op, val[_i]);
    }
  }

  op(82
  /* PushArgs */
  , names, blockNames, flags);
}

function SimpleArgs(op, positional, named, atNames) {
  if (positional === null && named === null) {
    op(83
    /* PushEmptyArgs */
    );
    return;
  }

  var count = CompilePositional(op, positional);
  var flags = count << 4;
  if (atNames) flags |= 8;
  var names = _util.EMPTY_STRING_ARRAY;

  if (named) {
    names = named[0];
    var val = named[1];

    for (var i = 0; i < val.length; i++) {
      (0, _expr.expr)(op, val[i]);
    }
  }

  op(82
  /* PushArgs */
  , names, _util.EMPTY_STRING_ARRAY, flags);
}
/**
 * Compile an optional list of positional arguments, which pushes each argument
 * onto the stack and returns the number of parameters compiled
 *
 * @param positional an optional list of positional arguments
 */


function CompilePositional(op, positional) {
  if (positional === null) return 0;

  for (var i = 0; i < positional.length; i++) {
    (0, _expr.expr)(op, positional[i]);
  }

  return positional.length;
}

function meta(layout) {
  var _a, _b;

  var _layout$block = layout.block,
      symbols = _layout$block[1],
      upvars = _layout$block[3];
  return {
    evalSymbols: evalSymbols(layout),
    upvars: upvars,
    scopeValues: (_b = (_a = layout.scope) === null || _a === void 0 ? void 0 : _a.call(layout)) !== null && _b !== void 0 ? _b : null,
    isStrictMode: layout.isStrictMode,
    moduleName: layout.moduleName,
    owner: layout.owner,
    size: symbols.length
  };
}

function evalSymbols(layout) {
  var block = layout.block;
  var symbols = block[1],
      hasEval = block[2];
  return hasEval ? symbols : null;
}
