// @glimmer/manager/dist/commonjs/es5/lib/internal/defaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FunctionHelperManager = void 0;

var _capabilities = require("../util/capabilities");

var FunctionHelperManager = /*#__PURE__*/function () {
  function FunctionHelperManager() {
    this.capabilities = (0, _capabilities.buildCapabilities)({
      hasValue: true,
      hasDestroyable: false,
      hasScheduledEffect: false
    });
  }

  var _proto = FunctionHelperManager.prototype;

  _proto.createHelper = function createHelper(fn, args) {
    return {
      fn: fn,
      args: args
    };
  };

  _proto.getValue = function getValue(_ref) {
    var fn = _ref.fn,
        args = _ref.args;

    if (Object.keys(args.named).length > 0) {
      var argsForFn = [].concat(args.positional, [args.named]);
      return fn.apply(void 0, argsForFn);
    }

    return fn.apply(void 0, args.positional);
  };

  _proto.getDebugName = function getDebugName(fn) {
    if (fn.name) {
      return "(helper function " + fn.name + ")";
    }

    return '(anonymous helper function)';
  };

  return FunctionHelperManager;
}();

exports.FunctionHelperManager = FunctionHelperManager;
