// @glimmer/compiler/dist/commonjs/es2017/lib/passes/1-normalization/keywords/append.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.APPEND_KEYWORDS = void 0;

var _syntax = require("@glimmer/syntax");

var _result = require("../../../shared/result");

var mir = _interopRequireWildcard(require("../../2-encoding/mir"));

var _expressions = require("../visitors/expressions");

var _impl = require("./impl");

var _callToAppend = require("./utils/call-to-append");

var _curry = require("./utils/curry");

var _dynamicVars = require("./utils/dynamic-vars");

var _hasBlock = require("./utils/has-block");

var _ifUnless = require("./utils/if-unless");

var _log = require("./utils/log");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const APPEND_KEYWORDS = (0, _impl.keywords)('Append').kw('has-block', (0, _callToAppend.toAppend)((0, _hasBlock.hasBlockKeyword)('has-block'))).kw('has-block-params', (0, _callToAppend.toAppend)((0, _hasBlock.hasBlockKeyword)('has-block-params'))).kw('-get-dynamic-var', (0, _callToAppend.toAppend)(_dynamicVars.getDynamicVarKeyword)).kw('log', (0, _callToAppend.toAppend)(_log.logKeyword)).kw('if', (0, _callToAppend.toAppend)((0, _ifUnless.ifUnlessInlineKeyword)('if'))).kw('unless', (0, _callToAppend.toAppend)((0, _ifUnless.ifUnlessInlineKeyword)('unless'))).kw('yield', {
  assert(node) {
    let {
      args
    } = node;

    if (args.named.isEmpty()) {
      return (0, _result.Ok)({
        target: _syntax.SourceSpan.synthetic('default').toSlice(),
        positional: args.positional
      });
    } else {
      let target = args.named.get('to');

      if (args.named.size > 1 || target === null) {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)(`yield only takes a single named argument: 'to'`, args.named.loc));
      }

      if (_syntax.ASTv2.isLiteral(target, 'string')) {
        return (0, _result.Ok)({
          target: target.toSlice(),
          positional: args.positional
        });
      } else {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)(`you can only yield to a literal string value`, target.loc));
      }
    }
  },

  translate({
    node,
    state
  }, {
    target,
    positional
  }) {
    return _expressions.VISIT_EXPRS.Positional(positional, state).mapOk(positional => new mir.Yield({
      loc: node.loc,
      target,
      to: state.scope.allocateBlock(target.chars),
      positional
    }));
  }

}).kw('debugger', {
  assert(node) {
    let {
      args
    } = node;
    let {
      positional
    } = args;

    if (args.isEmpty()) {
      return (0, _result.Ok)(undefined);
    } else {
      if (positional.isEmpty()) {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)(`debugger does not take any named arguments`, node.loc));
      } else {
        return (0, _result.Err)((0, _syntax.generateSyntaxError)(`debugger does not take any positional arguments`, node.loc));
      }
    }
  },

  translate({
    node,
    state: {
      scope
    }
  }) {
    scope.setHasEval();
    return (0, _result.Ok)(new mir.Debugger({
      loc: node.loc,
      scope
    }));
  }

}).kw('component', {
  assert: (0, _curry.assertCurryKeyword)(0
  /* Component */
  ),

  translate({
    node,
    state
  }, {
    definition,
    args
  }) {
    let definitionResult = _expressions.VISIT_EXPRS.visit(definition, state);

    let argsResult = _expressions.VISIT_EXPRS.Args(args, state);

    return _result.Result.all(definitionResult, argsResult).mapOk(([definition, args]) => new mir.InvokeComponent({
      loc: node.loc,
      definition,
      args,
      blocks: null
    }));
  }

}).kw('helper', {
  assert: (0, _curry.assertCurryKeyword)(1
  /* Helper */
  ),

  translate({
    node,
    state
  }, {
    definition,
    args
  }) {
    let definitionResult = _expressions.VISIT_EXPRS.visit(definition, state);

    let argsResult = _expressions.VISIT_EXPRS.Args(args, state);

    return _result.Result.all(definitionResult, argsResult).mapOk(([definition, args]) => {
      let text = new mir.CallExpression({
        callee: definition,
        args,
        loc: node.loc
      });
      return new mir.AppendTextNode({
        loc: node.loc,
        text
      });
    });
  }

});
exports.APPEND_KEYWORDS = APPEND_KEYWORDS;
