// @glimmer/compiler/dist/modules/es2017/lib/passes/1-normalization/visitors/element/simple-element.js
import { generateSyntaxError } from '@glimmer/syntax';
import { Err } from '../../../../shared/result';
import * as mir from '../../../2-encoding/mir';
import { VISIT_STMTS } from '../statements';
export class ClassifiedSimpleElement {
  constructor(tag, element, dynamicFeatures) {
    this.tag = tag;
    this.element = element;
    this.dynamicFeatures = dynamicFeatures;
    this.isComponent = false;
  }

  arg(attr) {
    return Err(generateSyntaxError(`${attr.name.chars} is not a valid attribute name. @arguments are only allowed on components, but the tag for this element (\`${this.tag.chars}\`) is a regular, non-component HTML element.`, attr.loc));
  }

  toStatement(classified, {
    params
  }) {
    let {
      state,
      element
    } = classified;
    let body = VISIT_STMTS.visitList(this.element.body, state);
    return body.mapOk(body => new mir.SimpleElement({
      loc: element.loc,
      tag: this.tag,
      params,
      body: body.toArray(),
      dynamicFeatures: this.dynamicFeatures
    }));
  }

}
