// @glimmer/compiler/dist/commonjs/es2017/lib/passes/1-normalization/keywords/utils/dynamic-vars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDynamicVarKeyword = void 0;

var _syntax = require("@glimmer/syntax");

var _result = require("../../../../shared/result");

var mir = _interopRequireWildcard(require("../../../2-encoding/mir"));

var _expressions = require("../../visitors/expressions");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function assertGetDynamicVarKeyword(node) {
  let call = node.type === 'AppendContent' ? node.value : node;
  let named = call.type === 'Call' ? call.args.named : null;
  let positionals = call.type === 'Call' ? call.args.positional : null;

  if (named && !named.isEmpty()) {
    return (0, _result.Err)((0, _syntax.generateSyntaxError)(`(-get-dynamic-vars) does not take any named arguments`, node.loc));
  }

  let varName = positionals === null || positionals === void 0 ? void 0 : positionals.nth(0);

  if (!varName) {
    return (0, _result.Err)((0, _syntax.generateSyntaxError)(`(-get-dynamic-vars) requires a var name to get`, node.loc));
  }

  if (positionals && positionals.size > 1) {
    return (0, _result.Err)((0, _syntax.generateSyntaxError)(`(-get-dynamic-vars) only receives one positional arg`, node.loc));
  }

  return (0, _result.Ok)(varName);
}

function translateGetDynamicVarKeyword({
  node,
  state
}, name) {
  return _expressions.VISIT_EXPRS.visit(name, state).mapOk(name => new mir.GetDynamicVar({
    name,
    loc: node.loc
  }));
}

const getDynamicVarKeyword = {
  assert: assertGetDynamicVarKeyword,
  translate: translateGetDynamicVarKeyword
};
exports.getDynamicVarKeyword = getDynamicVarKeyword;
