// @glimmer/runtime/dist/commonjs/es5/lib/vm/render-result.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _destroyable = require("@glimmer/destroyable");

var _bounds = require("../bounds");

var _update = _interopRequireDefault(require("./update"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RenderResultImpl = /*#__PURE__*/function () {
  function RenderResultImpl(env, updating, bounds, drop) {
    var _this = this;

    this.env = env;
    this.updating = updating;
    this.bounds = bounds;
    this.drop = drop;
    (0, _destroyable.associateDestroyableChild)(this, drop);
    (0, _destroyable.registerDestructor)(this, function () {
      return (0, _bounds.clear)(_this.bounds);
    });
  }

  var _proto = RenderResultImpl.prototype;

  _proto.rerender = function rerender(_temp) {
    var _ref = _temp === void 0 ? {
      alwaysRevalidate: false
    } : _temp,
        _ref$alwaysRevalidate = _ref.alwaysRevalidate,
        alwaysRevalidate = _ref$alwaysRevalidate === void 0 ? false : _ref$alwaysRevalidate;

    var env = this.env,
        updating = this.updating;
    var vm = new _update.default(env, {
      alwaysRevalidate: alwaysRevalidate
    });
    vm.execute(updating, this);
  };

  _proto.parentElement = function parentElement() {
    return this.bounds.parentElement();
  };

  _proto.firstNode = function firstNode() {
    return this.bounds.firstNode();
  };

  _proto.lastNode = function lastNode() {
    return this.bounds.lastNode();
  };

  _proto.handleException = function handleException() {
    throw 'this should never happen';
  };

  return RenderResultImpl;
}();

exports.default = RenderResultImpl;
