// @glimmer/runtime/dist/commonjs/es5/lib/vm/stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vm = require("@glimmer/vm");

var _symbols = require("../symbols");

var _lowLevel = require("./low-level");

var EvaluationStackImpl = /*#__PURE__*/function () {
  // fp -> sp
  function EvaluationStackImpl(stack, registers) {
    if (stack === void 0) {
      stack = [];
    }

    this.stack = stack;
    this[_symbols.REGISTERS] = registers;

    if (false
    /* LOCAL_DEBUG */
    ) {
        Object.seal(this);
      }
  }

  EvaluationStackImpl.restore = function restore(snapshot) {
    return new this(snapshot.slice(), (0, _lowLevel.initializeRegistersWithSP)(snapshot.length - 1));
  };

  var _proto = EvaluationStackImpl.prototype;

  _proto.push = function push(value) {
    this.stack[++this[_symbols.REGISTERS][_vm.$sp]] = value;
  };

  _proto.dup = function dup(position) {
    if (position === void 0) {
      position = this[_symbols.REGISTERS][_vm.$sp];
    }

    this.stack[++this[_symbols.REGISTERS][_vm.$sp]] = this.stack[position];
  };

  _proto.copy = function copy(from, to) {
    this.stack[to] = this.stack[from];
  };

  _proto.pop = function pop(n) {
    if (n === void 0) {
      n = 1;
    }

    var top = this.stack[this[_symbols.REGISTERS][_vm.$sp]];
    this[_symbols.REGISTERS][_vm.$sp] -= n;
    return top;
  };

  _proto.peek = function peek(offset) {
    if (offset === void 0) {
      offset = 0;
    }

    return this.stack[this[_symbols.REGISTERS][_vm.$sp] - offset];
  };

  _proto.get = function get(offset, base) {
    if (base === void 0) {
      base = this[_symbols.REGISTERS][_vm.$fp];
    }

    return this.stack[base + offset];
  };

  _proto.set = function set(value, offset, base) {
    if (base === void 0) {
      base = this[_symbols.REGISTERS][_vm.$fp];
    }

    this.stack[base + offset] = value;
  };

  _proto.slice = function slice(start, end) {
    return this.stack.slice(start, end);
  };

  _proto.capture = function capture(items) {
    var end = this[_symbols.REGISTERS][_vm.$sp] + 1;
    var start = end - items;
    return this.stack.slice(start, end);
  };

  _proto.reset = function reset() {
    this.stack.length = 0;
  };

  _proto.toArray = function toArray() {
    return this.stack.slice(this[_symbols.REGISTERS][_vm.$fp], this[_symbols.REGISTERS][_vm.$sp] + 1);
  };

  return EvaluationStackImpl;
}();

exports.default = EvaluationStackImpl;
