// @glimmer/runtime/dist/commonjs/es5/lib/component/template-only.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.templateOnlyComponent = templateOnlyComponent;
exports.TemplateOnlyComponentDefinition = exports.TEMPLATE_ONLY_COMPONENT_MANAGER = exports.TemplateOnlyComponentManager = void 0;

var _reference = require("@glimmer/reference");

var _manager = require("@glimmer/manager");

var CAPABILITIES = {
  dynamicLayout: false,
  dynamicTag: false,
  prepareArgs: false,
  createArgs: false,
  attributeHook: false,
  elementHook: false,
  createCaller: false,
  dynamicScope: false,
  updateHook: false,
  createInstance: false,
  wrapped: false,
  willDestroy: false,
  hasSubOwner: false
};

var TemplateOnlyComponentManager = /*#__PURE__*/function () {
  function TemplateOnlyComponentManager() {}

  var _proto = TemplateOnlyComponentManager.prototype;

  _proto.getCapabilities = function getCapabilities() {
    return CAPABILITIES;
  };

  _proto.getDebugName = function getDebugName(_ref) {
    var name = _ref.name;
    return name;
  };

  _proto.getSelf = function getSelf() {
    return _reference.NULL_REFERENCE;
  };

  _proto.getDestroyable = function getDestroyable() {
    return null;
  };

  return TemplateOnlyComponentManager;
}();

exports.TemplateOnlyComponentManager = TemplateOnlyComponentManager;
var TEMPLATE_ONLY_COMPONENT_MANAGER = new TemplateOnlyComponentManager(); // This is only exported for types, don't use this class directly

exports.TEMPLATE_ONLY_COMPONENT_MANAGER = TEMPLATE_ONLY_COMPONENT_MANAGER;

var TemplateOnlyComponentDefinition = /*#__PURE__*/function () {
  function TemplateOnlyComponentDefinition(moduleName, name) {
    if (moduleName === void 0) {
      moduleName = '@glimmer/component/template-only';
    }

    if (name === void 0) {
      name = '(unknown template-only component)';
    }

    this.moduleName = moduleName;
    this.name = name;
  }

  var _proto2 = TemplateOnlyComponentDefinition.prototype;

  _proto2.toString = function toString() {
    return this.moduleName;
  };

  return TemplateOnlyComponentDefinition;
}();

exports.TemplateOnlyComponentDefinition = TemplateOnlyComponentDefinition;
(0, _manager.setInternalComponentManager)(TEMPLATE_ONLY_COMPONENT_MANAGER, TemplateOnlyComponentDefinition.prototype);
/**
  This utility function is used to declare a given component has no backing class. When the rendering engine detects this it
  is able to perform a number of optimizations. Templates that are associated with `templateOnly()` will be rendered _as is_
  without adding a wrapping `<div>` (or any of the other element customization behaviors of [@ember/component](/ember/release/classes/Component)).
  Specifically, this means that the template will be rendered as "outer HTML".

  In general, this method will be used by build time tooling and would not be directly written in an application. However,
  at times it may be useful to use directly to leverage the "outer HTML" semantics mentioned above. For example, if an addon would like
  to use these semantics for its templates but cannot be certain it will only be consumed by applications that have enabled the
  `template-only-glimmer-components` optional feature.

  @example

  ```js
  import { templateOnlyComponent } from '@glimmer/runtime';

  export default templateOnlyComponent();
  ```

  @public
  @method templateOnly
  @param {String} moduleName the module name that the template only component represents, this will be used for debugging purposes
  @category EMBER_GLIMMER_SET_COMPONENT_TEMPLATE
*/

function templateOnlyComponent(moduleName, name) {
  return new TemplateOnlyComponentDefinition(moduleName, name);
}
