// @glimmer/compiler/dist/commonjs/es2017/lib/passes/1-normalization/visitors/expressions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.convertPathToCallIfKeyword = convertPathToCallIfKeyword;
exports.VISIT_EXPRS = exports.NormalizeExpressions = void 0;

var _syntax = require("@glimmer/syntax");

var _util = require("@glimmer/util");

var _result = require("../../../shared/result");

var mir = _interopRequireWildcard(require("../../2-encoding/mir"));

var _keywords = require("../keywords");

var _isNode = require("../utils/is-node");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

class NormalizeExpressions {
  visit(node, state) {
    switch (node.type) {
      case 'Literal':
        return (0, _result.Ok)(this.Literal(node));

      case 'Interpolate':
        return this.Interpolate(node, state);

      case 'Path':
        return this.PathExpression(node);

      case 'Call':
        let translated = _keywords.CALL_KEYWORDS.translate(node, state);

        if (translated !== null) {
          return translated;
        }

        return this.CallExpression(node, state);

      case 'DeprecatedCall':
        return this.DeprecaedCallExpression(node, state);
    }
  }

  visitList(nodes, state) {
    return new _result.ResultArray(nodes.map(e => VISIT_EXPRS.visit(e, state))).toOptionalList();
  }
  /**
   * Normalize paths into `hir.Path` or a `hir.Expr` that corresponds to the ref.
   *
   * TODO since keywords don't support tails anyway, distinguish PathExpression from
   * VariableReference in ASTv2.
   */


  PathExpression(path) {
    let ref = this.VariableReference(path.ref);
    let {
      tail
    } = path;

    if ((0, _util.isPresent)(tail)) {
      let tailLoc = tail[0].loc.extend(tail[tail.length - 1].loc);
      return (0, _result.Ok)(new mir.PathExpression({
        loc: path.loc,
        head: ref,
        tail: new mir.Tail({
          loc: tailLoc,
          members: tail
        })
      }));
    } else {
      return (0, _result.Ok)(ref);
    }
  }

  VariableReference(ref) {
    return ref;
  }

  Literal(literal) {
    return literal;
  }

  Interpolate(expr, state) {
    let parts = expr.parts.map(convertPathToCallIfKeyword);
    return VISIT_EXPRS.visitList(parts, state).mapOk(parts => new mir.InterpolateExpression({
      loc: expr.loc,
      parts: parts
    }));
  }

  CallExpression(expr, state) {
    if (!(0, _isNode.hasPath)(expr)) {
      throw new Error(`unimplemented subexpression at the head of a subexpression`);
    } else {
      return _result.Result.all(VISIT_EXPRS.visit(expr.callee, state), VISIT_EXPRS.Args(expr.args, state)).mapOk(([callee, args]) => new mir.CallExpression({
        loc: expr.loc,
        callee,
        args
      }));
    }
  }

  DeprecaedCallExpression({
    arg,
    callee,
    loc
  }, _state) {
    return (0, _result.Ok)(new mir.DeprecatedCallExpression({
      loc,
      arg,
      callee
    }));
  }

  Args({
    positional,
    named,
    loc
  }, state) {
    return _result.Result.all(this.Positional(positional, state), this.NamedArguments(named, state)).mapOk(([positional, named]) => new mir.Args({
      loc,
      positional,
      named
    }));
  }

  Positional(positional, state) {
    return VISIT_EXPRS.visitList(positional.exprs, state).mapOk(list => new mir.Positional({
      loc: positional.loc,
      list
    }));
  }

  NamedArguments(named, state) {
    let pairs = named.entries.map(arg => {
      let value = convertPathToCallIfKeyword(arg.value);
      return VISIT_EXPRS.visit(value, state).mapOk(value => new mir.NamedArgument({
        loc: arg.loc,
        key: arg.name,
        value
      }));
    });
    return new _result.ResultArray(pairs).toOptionalList().mapOk(pairs => new mir.NamedArguments({
      loc: named.loc,
      entries: pairs
    }));
  }

}

exports.NormalizeExpressions = NormalizeExpressions;

function convertPathToCallIfKeyword(path) {
  if (path.type === 'Path' && path.ref.type === 'Free' && path.ref.name in _syntax.KEYWORDS_TYPES) {
    return new _syntax.ASTv2.CallExpression({
      callee: path,
      args: _syntax.ASTv2.Args.empty(path.loc),
      loc: path.loc
    });
  }

  return path;
}

const VISIT_EXPRS = new NormalizeExpressions();
exports.VISIT_EXPRS = VISIT_EXPRS;
