// @glimmer/opcode-compiler/dist/commonjs/es2017/lib/compilable-template.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compilable = compilable;
exports.compileStatements = compileStatements;
exports.compilableBlock = compilableBlock;
exports.PLACEHOLDER_HANDLE = void 0;

var _shared = require("./opcode-builder/helpers/shared");

var _util = require("@glimmer/util");

var _context = require("./opcode-builder/context");

var _compiler = require("./compiler");

var _statements = require("./syntax/statements");

var _encoder = require("./opcode-builder/encoder");

const PLACEHOLDER_HANDLE = -1;
exports.PLACEHOLDER_HANDLE = PLACEHOLDER_HANDLE;

class CompilableTemplateImpl {
  constructor(statements, meta, // Part of CompilableTemplate
  symbolTable, // Used for debugging
  moduleName = 'plain block') {
    this.statements = statements;
    this.meta = meta;
    this.symbolTable = symbolTable;
    this.moduleName = moduleName;
    this.compiled = null;
  } // Part of CompilableTemplate


  compile(context) {
    return maybeCompile(this, context);
  }

}

function compilable(layout, moduleName) {
  let [statements, symbols, hasEval] = layout.block;
  return new CompilableTemplateImpl(statements, (0, _shared.meta)(layout), {
    symbols,
    hasEval
  }, moduleName);
}

function maybeCompile(compilable, context) {
  if (compilable.compiled !== null) return compilable.compiled;
  compilable.compiled = PLACEHOLDER_HANDLE;
  let {
    statements,
    meta
  } = compilable;
  let result = compileStatements(statements, meta, context);
  compilable.compiled = result;
  return result;
}

function compileStatements(statements, meta, syntaxContext) {
  let sCompiler = _statements.STATEMENTS;
  let context = (0, _context.templateCompilationContext)(syntaxContext, meta);
  let {
    encoder,
    program: {
      constants,
      resolver
    }
  } = context;

  function pushOp(...op) {
    (0, _encoder.encodeOp)(encoder, constants, resolver, meta, op);
  }

  for (let i = 0; i < statements.length; i++) {
    sCompiler.compile(pushOp, statements[i]);
  }

  let handle = context.encoder.commit(meta.size);

  if (false
  /* LOCAL_SHOULD_LOG */
  ) {
      (0, _compiler.debugCompiler)(context, handle);
    }

  return handle;
}

function compilableBlock(block, containing) {
  return new CompilableTemplateImpl(block[0], containing, {
    parameters: block[1] || _util.EMPTY_ARRAY
  });
}
