// @glimmer/opcode-compiler/dist/commonjs/es2017/lib/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.namedBlocks = namedBlocks;
exports.EMPTY_BLOCKS = exports.NamedBlocksImpl = void 0;

var _util = require("@glimmer/util");

class NamedBlocksImpl {
  constructor(blocks) {
    this.blocks = blocks;
    this.names = blocks ? Object.keys(blocks) : [];
  }

  get(name) {
    if (!this.blocks) return null;
    return this.blocks[name] || null;
  }

  has(name) {
    let {
      blocks
    } = this;
    return blocks !== null && name in blocks;
  }

  with(name, block) {
    let {
      blocks
    } = this;

    if (blocks) {
      return new NamedBlocksImpl((0, _util.assign)({}, blocks, {
        [name]: block
      }));
    } else {
      return new NamedBlocksImpl({
        [name]: block
      });
    }
  }

  get hasAny() {
    return this.blocks !== null;
  }

}

exports.NamedBlocksImpl = NamedBlocksImpl;
const EMPTY_BLOCKS = new NamedBlocksImpl(null);
exports.EMPTY_BLOCKS = EMPTY_BLOCKS;

function namedBlocks(blocks) {
  if (blocks === null) {
    return EMPTY_BLOCKS;
  }

  let out = (0, _util.dict)();
  let [keys, values] = blocks;

  for (let i = 0; i < keys.length; i++) {
    out[keys[i]] = values[i];
  }

  return new NamedBlocksImpl(out);
}
