// @glimmer/opcode-compiler/dist/commonjs/es2017/lib/wrapped-component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WrappedBuilder = void 0;

var _context = require("./opcode-builder/context");

var _shared = require("./opcode-builder/helpers/shared");

var _components = require("./opcode-builder/helpers/components");

var _compiler = require("./compiler");

var _encoder = require("./opcode-builder/encoder");

class WrappedBuilder {
  constructor(layout, moduleName) {
    this.layout = layout;
    this.moduleName = moduleName;
    this.compiled = null;
    let {
      block
    } = layout;
    let [, symbols, hasEval] = block;
    symbols = symbols.slice(); // ensure ATTRS_BLOCK is always included (only once) in the list of symbols

    let attrsBlockIndex = symbols.indexOf(_components.ATTRS_BLOCK);

    if (attrsBlockIndex === -1) {
      this.attrsBlockNumber = symbols.push(_components.ATTRS_BLOCK);
    } else {
      this.attrsBlockNumber = attrsBlockIndex + 1;
    }

    this.symbolTable = {
      hasEval,
      symbols
    };
  }

  compile(syntax) {
    if (this.compiled !== null) return this.compiled;
    let m = (0, _shared.meta)(this.layout);
    let context = (0, _context.templateCompilationContext)(syntax, m);
    let {
      encoder,
      program: {
        constants,
        resolver
      }
    } = context;

    function pushOp(...op) {
      (0, _encoder.encodeOp)(encoder, constants, resolver, m, op);
    }

    (0, _components.WrappedComponent)(pushOp, this.layout, this.attrsBlockNumber);
    let handle = context.encoder.commit(m.size);

    if (typeof handle !== 'number') {
      return handle;
    }

    this.compiled = handle;

    if (false
    /* LOCAL_SHOULD_LOG */
    ) {
        (0, _compiler.debugCompiler)(context, handle);
      }

    return handle;
  }

}

exports.WrappedBuilder = WrappedBuilder;
