// @glimmer/compiler/dist/modules/es5/lib/passes/2-encoding/mir.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { node } from '@glimmer/syntax';
export var Template = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(Template, _node$fields);

  function Template() {
    return _node$fields.apply(this, arguments) || this;
  }

  return Template;
}(node('Template').fields());
export var InElement = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(InElement, _node$fields2);

  function InElement() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return InElement;
}(node('InElement').fields());
export var Not = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(Not, _node$fields3);

  function Not() {
    return _node$fields3.apply(this, arguments) || this;
  }

  return Not;
}(node('Not').fields());
export var If = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(If, _node$fields4);

  function If() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return If;
}(node('If').fields());
export var IfInline = /*#__PURE__*/function (_node$fields5) {
  _inheritsLoose(IfInline, _node$fields5);

  function IfInline() {
    return _node$fields5.apply(this, arguments) || this;
  }

  return IfInline;
}(node('IfInline').fields());
export var Each = /*#__PURE__*/function (_node$fields6) {
  _inheritsLoose(Each, _node$fields6);

  function Each() {
    return _node$fields6.apply(this, arguments) || this;
  }

  return Each;
}(node('Each').fields());
export var With = /*#__PURE__*/function (_node$fields7) {
  _inheritsLoose(With, _node$fields7);

  function With() {
    return _node$fields7.apply(this, arguments) || this;
  }

  return With;
}(node('With').fields());
export var Let = /*#__PURE__*/function (_node$fields8) {
  _inheritsLoose(Let, _node$fields8);

  function Let() {
    return _node$fields8.apply(this, arguments) || this;
  }

  return Let;
}(node('Let').fields());
export var WithDynamicVars = /*#__PURE__*/function (_node$fields9) {
  _inheritsLoose(WithDynamicVars, _node$fields9);

  function WithDynamicVars() {
    return _node$fields9.apply(this, arguments) || this;
  }

  return WithDynamicVars;
}(node('WithDynamicVars').fields());
export var GetDynamicVar = /*#__PURE__*/function (_node$fields10) {
  _inheritsLoose(GetDynamicVar, _node$fields10);

  function GetDynamicVar() {
    return _node$fields10.apply(this, arguments) || this;
  }

  return GetDynamicVar;
}(node('GetDynamicVar').fields());
export var Log = /*#__PURE__*/function (_node$fields11) {
  _inheritsLoose(Log, _node$fields11);

  function Log() {
    return _node$fields11.apply(this, arguments) || this;
  }

  return Log;
}(node('Log').fields());
export var InvokeComponent = /*#__PURE__*/function (_node$fields12) {
  _inheritsLoose(InvokeComponent, _node$fields12);

  function InvokeComponent() {
    return _node$fields12.apply(this, arguments) || this;
  }

  return InvokeComponent;
}(node('InvokeComponent').fields());
export var NamedBlocks = /*#__PURE__*/function (_node$fields13) {
  _inheritsLoose(NamedBlocks, _node$fields13);

  function NamedBlocks() {
    return _node$fields13.apply(this, arguments) || this;
  }

  return NamedBlocks;
}(node('NamedBlocks').fields());
export var NamedBlock = /*#__PURE__*/function (_node$fields14) {
  _inheritsLoose(NamedBlock, _node$fields14);

  function NamedBlock() {
    return _node$fields14.apply(this, arguments) || this;
  }

  return NamedBlock;
}(node('NamedBlock').fields());
export var EndBlock = /*#__PURE__*/function (_node$fields15) {
  _inheritsLoose(EndBlock, _node$fields15);

  function EndBlock() {
    return _node$fields15.apply(this, arguments) || this;
  }

  return EndBlock;
}(node('EndBlock').fields());
export var AppendTrustedHTML = /*#__PURE__*/function (_node$fields16) {
  _inheritsLoose(AppendTrustedHTML, _node$fields16);

  function AppendTrustedHTML() {
    return _node$fields16.apply(this, arguments) || this;
  }

  return AppendTrustedHTML;
}(node('AppendTrustedHTML').fields());
export var AppendTextNode = /*#__PURE__*/function (_node$fields17) {
  _inheritsLoose(AppendTextNode, _node$fields17);

  function AppendTextNode() {
    return _node$fields17.apply(this, arguments) || this;
  }

  return AppendTextNode;
}(node('AppendTextNode').fields());
export var AppendComment = /*#__PURE__*/function (_node$fields18) {
  _inheritsLoose(AppendComment, _node$fields18);

  function AppendComment() {
    return _node$fields18.apply(this, arguments) || this;
  }

  return AppendComment;
}(node('AppendComment').fields());
export var Component = /*#__PURE__*/function (_node$fields19) {
  _inheritsLoose(Component, _node$fields19);

  function Component() {
    return _node$fields19.apply(this, arguments) || this;
  }

  return Component;
}(node('Component').fields());
export var StaticAttr = /*#__PURE__*/function (_node$fields20) {
  _inheritsLoose(StaticAttr, _node$fields20);

  function StaticAttr() {
    return _node$fields20.apply(this, arguments) || this;
  }

  return StaticAttr;
}(node('StaticAttr').fields());
export var DynamicAttr = /*#__PURE__*/function (_node$fields21) {
  _inheritsLoose(DynamicAttr, _node$fields21);

  function DynamicAttr() {
    return _node$fields21.apply(this, arguments) || this;
  }

  return DynamicAttr;
}(node('DynamicAttr').fields());
export var SimpleElement = /*#__PURE__*/function (_node$fields22) {
  _inheritsLoose(SimpleElement, _node$fields22);

  function SimpleElement() {
    return _node$fields22.apply(this, arguments) || this;
  }

  return SimpleElement;
}(node('SimpleElement').fields());
export var ElementParameters = /*#__PURE__*/function (_node$fields23) {
  _inheritsLoose(ElementParameters, _node$fields23);

  function ElementParameters() {
    return _node$fields23.apply(this, arguments) || this;
  }

  return ElementParameters;
}(node('ElementParameters').fields());
export var Yield = /*#__PURE__*/function (_node$fields24) {
  _inheritsLoose(Yield, _node$fields24);

  function Yield() {
    return _node$fields24.apply(this, arguments) || this;
  }

  return Yield;
}(node('Yield').fields());
export var Debugger = /*#__PURE__*/function (_node$fields25) {
  _inheritsLoose(Debugger, _node$fields25);

  function Debugger() {
    return _node$fields25.apply(this, arguments) || this;
  }

  return Debugger;
}(node('Debugger').fields());
export var CallExpression = /*#__PURE__*/function (_node$fields26) {
  _inheritsLoose(CallExpression, _node$fields26);

  function CallExpression() {
    return _node$fields26.apply(this, arguments) || this;
  }

  return CallExpression;
}(node('CallExpression').fields());
export var DeprecatedCallExpression = /*#__PURE__*/function (_node$fields27) {
  _inheritsLoose(DeprecatedCallExpression, _node$fields27);

  function DeprecatedCallExpression() {
    return _node$fields27.apply(this, arguments) || this;
  }

  return DeprecatedCallExpression;
}(node('DeprecatedCallExpression').fields());
export var Modifier = /*#__PURE__*/function (_node$fields28) {
  _inheritsLoose(Modifier, _node$fields28);

  function Modifier() {
    return _node$fields28.apply(this, arguments) || this;
  }

  return Modifier;
}(node('Modifier').fields());
export var InvokeBlock = /*#__PURE__*/function (_node$fields29) {
  _inheritsLoose(InvokeBlock, _node$fields29);

  function InvokeBlock() {
    return _node$fields29.apply(this, arguments) || this;
  }

  return InvokeBlock;
}(node('InvokeBlock').fields());
export var SplatAttr = /*#__PURE__*/function (_node$fields30) {
  _inheritsLoose(SplatAttr, _node$fields30);

  function SplatAttr() {
    return _node$fields30.apply(this, arguments) || this;
  }

  return SplatAttr;
}(node('SplatAttr').fields());
export var PathExpression = /*#__PURE__*/function (_node$fields31) {
  _inheritsLoose(PathExpression, _node$fields31);

  function PathExpression() {
    return _node$fields31.apply(this, arguments) || this;
  }

  return PathExpression;
}(node('PathExpression').fields());
export var GetWithResolver = /*#__PURE__*/function (_node$fields32) {
  _inheritsLoose(GetWithResolver, _node$fields32);

  function GetWithResolver() {
    return _node$fields32.apply(this, arguments) || this;
  }

  return GetWithResolver;
}(node('GetWithResolver').fields());
export var GetSymbol = /*#__PURE__*/function (_node$fields33) {
  _inheritsLoose(GetSymbol, _node$fields33);

  function GetSymbol() {
    return _node$fields33.apply(this, arguments) || this;
  }

  return GetSymbol;
}(node('GetSymbol').fields());
export var GetFreeWithContext = /*#__PURE__*/function (_node$fields34) {
  _inheritsLoose(GetFreeWithContext, _node$fields34);

  function GetFreeWithContext() {
    return _node$fields34.apply(this, arguments) || this;
  }

  return GetFreeWithContext;
}(node('GetFreeWithContext').fields());
/** strict mode */

export var GetFree = /*#__PURE__*/function (_node$fields35) {
  _inheritsLoose(GetFree, _node$fields35);

  function GetFree() {
    return _node$fields35.apply(this, arguments) || this;
  }

  return GetFree;
}(node('GetFree').fields());
export var Missing = /*#__PURE__*/function (_node$fields36) {
  _inheritsLoose(Missing, _node$fields36);

  function Missing() {
    return _node$fields36.apply(this, arguments) || this;
  }

  return Missing;
}(node('Missing').fields());
export var InterpolateExpression = /*#__PURE__*/function (_node$fields37) {
  _inheritsLoose(InterpolateExpression, _node$fields37);

  function InterpolateExpression() {
    return _node$fields37.apply(this, arguments) || this;
  }

  return InterpolateExpression;
}(node('InterpolateExpression').fields());
export var HasBlock = /*#__PURE__*/function (_node$fields38) {
  _inheritsLoose(HasBlock, _node$fields38);

  function HasBlock() {
    return _node$fields38.apply(this, arguments) || this;
  }

  return HasBlock;
}(node('HasBlock').fields());
export var HasBlockParams = /*#__PURE__*/function (_node$fields39) {
  _inheritsLoose(HasBlockParams, _node$fields39);

  function HasBlockParams() {
    return _node$fields39.apply(this, arguments) || this;
  }

  return HasBlockParams;
}(node('HasBlockParams').fields());
export var Curry = /*#__PURE__*/function (_node$fields40) {
  _inheritsLoose(Curry, _node$fields40);

  function Curry() {
    return _node$fields40.apply(this, arguments) || this;
  }

  return Curry;
}(node('Curry').fields());
export var Positional = /*#__PURE__*/function (_node$fields41) {
  _inheritsLoose(Positional, _node$fields41);

  function Positional() {
    return _node$fields41.apply(this, arguments) || this;
  }

  return Positional;
}(node('Positional').fields());
export var NamedArguments = /*#__PURE__*/function (_node$fields42) {
  _inheritsLoose(NamedArguments, _node$fields42);

  function NamedArguments() {
    return _node$fields42.apply(this, arguments) || this;
  }

  return NamedArguments;
}(node('NamedArguments').fields());
export var NamedArgument = /*#__PURE__*/function (_node$fields43) {
  _inheritsLoose(NamedArgument, _node$fields43);

  function NamedArgument() {
    return _node$fields43.apply(this, arguments) || this;
  }

  return NamedArgument;
}(node('NamedArgument').fields());
export var Args = /*#__PURE__*/function (_node$fields44) {
  _inheritsLoose(Args, _node$fields44);

  function Args() {
    return _node$fields44.apply(this, arguments) || this;
  }

  return Args;
}(node('Args').fields());
export var Tail = /*#__PURE__*/function (_node$fields45) {
  _inheritsLoose(Tail, _node$fields45);

  function Tail() {
    return _node$fields45.apply(this, arguments) || this;
  }

  return Tail;
}(node('Tail').fields());
