// @glimmer/compiler/dist/commonjs/es5/lib/shared/list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OptionalList = OptionalList;
exports.EmptyList = exports.PresentList = void 0;

var _util = require("@glimmer/util");

function _createForOfIteratorHelperLoose(o, allowArrayLike) {
  var it;

  if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) {
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;
      return function () {
        if (i >= o.length) return {
          done: true
        };
        return {
          done: false,
          value: o[i++]
        };
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  it = o[Symbol.iterator]();
  return it.next.bind(it);
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var PresentList = /*#__PURE__*/function () {
  function PresentList(list) {
    this.list = list;
  }

  var _proto = PresentList.prototype;

  _proto.toArray = function toArray() {
    return this.list;
  };

  _proto.map = function map(callback) {
    var result = (0, _util.mapPresent)(this.list, callback);
    return new PresentList(result);
  };

  _proto.filter = function filter(predicate) {
    var out = [];

    for (var _iterator = _createForOfIteratorHelperLoose(this.list), _step; !(_step = _iterator()).done;) {
      var item = _step.value;

      if (predicate(item)) {
        out.push(item);
      }
    }

    return OptionalList(out);
  };

  _proto.toPresentArray = function toPresentArray() {
    return this.list;
  };

  _proto.into = function into(_ref) {
    var ifPresent = _ref.ifPresent;
    return ifPresent(this);
  };

  return PresentList;
}();

exports.PresentList = PresentList;

var EmptyList = /*#__PURE__*/function () {
  function EmptyList() {
    this.list = [];
  }

  var _proto2 = EmptyList.prototype;

  _proto2.map = function map(_callback) {
    return new EmptyList();
  };

  _proto2.filter = function filter(_predicate) {
    return new EmptyList();
  };

  _proto2.toArray = function toArray() {
    return this.list;
  };

  _proto2.toPresentArray = function toPresentArray() {
    return null;
  };

  _proto2.into = function into(_ref2) {
    var ifEmpty = _ref2.ifEmpty;
    return ifEmpty();
  };

  return EmptyList;
}(); // export type OptionalList<T> = PresentList<T> | EmptyList<T>;


exports.EmptyList = EmptyList;

function OptionalList(value) {
  if ((0, _util.isPresent)(value)) {
    return new PresentList(value);
  } else {
    return new EmptyList();
  }
}
