// @glimmer/opcode-compiler/dist/commonjs/es2017/lib/opcode-builder/operands.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.labelOperand = labelOperand;
exports.evalSymbolsOperand = evalSymbolsOperand;
exports.isStrictMode = isStrictMode;
exports.blockOperand = blockOperand;
exports.stdlibOperand = stdlibOperand;
exports.nonSmallIntOperand = nonSmallIntOperand;
exports.symbolTableOperand = symbolTableOperand;
exports.layoutOperand = layoutOperand;

var _util = require("@glimmer/util");

function labelOperand(value) {
  return {
    type: 1
    /* Label */
    ,
    value
  };
}

function evalSymbolsOperand() {
  return {
    type: 3
    /* EvalSymbols */
    ,
    value: undefined
  };
}

function isStrictMode() {
  return {
    type: 2
    /* IsStrictMode */
    ,
    value: undefined
  };
}

function blockOperand(value) {
  return {
    type: 4
    /* Block */
    ,
    value
  };
}

function stdlibOperand(value) {
  return {
    type: 5
    /* StdLib */
    ,
    value
  };
}

function nonSmallIntOperand(value) {
  false && (0, _util.assert)(!(0, _util.isSmallInt)(value), 'Attempted to make a operand for an int that was not a small int, you should encode this as an immediate');
  return {
    type: 6
    /* NonSmallInt */
    ,
    value
  };
}

function symbolTableOperand(value) {
  return {
    type: 7
    /* SymbolTable */
    ,
    value
  };
}

function layoutOperand(value) {
  return {
    type: 8
    /* Layout */
    ,
    value
  };
}
