// @glimmer/opcode-compiler/dist/commonjs/es2017/lib/opcode-builder/helpers/vm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PushPrimitiveReference = PushPrimitiveReference;
exports.PushPrimitive = PushPrimitive;
exports.Call = Call;
exports.CallDynamic = CallDynamic;
exports.DynamicScope = DynamicScope;
exports.Curry = Curry;

var _vm = require("@glimmer/vm");

var _util = require("@glimmer/util");

var _shared = require("./shared");

var _operands = require("../operands");

var _expr = require("./expr");

/**
 * Push a reference onto the stack corresponding to a statically known primitive
 * @param value A JavaScript primitive (undefined, null, boolean, number or string)
 */
function PushPrimitiveReference(op, value) {
  PushPrimitive(op, value);
  op(31
  /* PrimitiveReference */
  );
}
/**
 * Push an encoded representation of a JavaScript primitive on the stack
 *
 * @param value A JavaScript primitive (undefined, null, boolean, number or string)
 */


function PushPrimitive(op, primitive) {
  let p = primitive;

  if (typeof p === 'number') {
    p = (0, _util.isSmallInt)(p) ? (0, _util.encodeImmediate)(p) : (0, _operands.nonSmallIntOperand)(p);
  }

  op(30
  /* Primitive */
  , p);
}
/**
 * Invoke a foreign function (a "helper") based on a statically known handle
 *
 * @param op The op creation function
 * @param handle A handle
 * @param positional An optional list of expressions to compile
 * @param named An optional list of named arguments (name + expression) to compile
 */


function Call(op, handle, positional, named) {
  op(0
  /* PushFrame */
  );
  (0, _shared.SimpleArgs)(op, positional, named, false);
  op(16
  /* Helper */
  , handle);
  op(1
  /* PopFrame */
  );
  op(36
  /* Fetch */
  , _vm.$v0);
}
/**
 * Invoke a foreign function (a "helper") based on a dynamically loaded definition
 *
 * @param op The op creation function
 * @param positional An optional list of expressions to compile
 * @param named An optional list of named arguments (name + expression) to compile
 */


function CallDynamic(op, positional, named, append) {
  op(0
  /* PushFrame */
  );
  (0, _shared.SimpleArgs)(op, positional, named, false);
  op(33
  /* Dup */
  , _vm.$fp, 1);
  op(107
  /* DynamicHelper */
  );

  if (append) {
    op(36
    /* Fetch */
    , _vm.$v0);
    append();
    op(1
    /* PopFrame */
    );
    op(34
    /* Pop */
    , 1);
  } else {
    op(1
    /* PopFrame */
    );
    op(34
    /* Pop */
    , 1);
    op(36
    /* Fetch */
    , _vm.$v0);
  }
}
/**
 * Evaluate statements in the context of new dynamic scope entries. Move entries from the
 * stack into named entries in the dynamic scope, then evaluate the statements, then pop
 * the dynamic scope
 *
 * @param names a list of dynamic scope names
 * @param block a function that returns a list of statements to evaluate
 */


function DynamicScope(op, names, block) {
  op(59
  /* PushDynamicScope */
  );
  op(58
  /* BindDynamicScope */
  , names);
  block();
  op(60
  /* PopDynamicScope */
  );
}

function Curry(op, type, definition, positional, named) {
  op(0
  /* PushFrame */
  );
  (0, _shared.SimpleArgs)(op, positional, named, false);
  op(86
  /* CaptureArgs */
  );
  (0, _expr.expr)(op, definition);
  op(77
  /* Curry */
  , type, (0, _operands.isStrictMode)());
  op(1
  /* PopFrame */
  );
  op(36
  /* Fetch */
  , _vm.$v0);
}
