// @glimmer/runtime/dist/modules/es2017/lib/dom/normalize.js
export function normalizeStringValue(value) {
  if (isEmpty(value)) {
    return '';
  }

  return String(value);
}
export function normalizeTrustedValue(value) {
  if (isEmpty(value)) {
    return '';
  }

  if (isString(value)) {
    return value;
  }

  if (isSafeString(value)) {
    return value.toHTML();
  }

  if (isNode(value)) {
    return value;
  }

  return String(value);
}
export function shouldCoerce(value) {
  return isString(value) || isEmpty(value) || typeof value === 'boolean' || typeof value === 'number';
}
export function isEmpty(value) {
  return value === null || value === undefined || typeof value.toString !== 'function';
}
export function isSafeString(value) {
  return typeof value === 'object' && value !== null && typeof value.toHTML === 'function';
}
export function isNode(value) {
  return typeof value === 'object' && value !== null && typeof value.nodeType === 'number';
}
export function isFragment(value) {
  return isNode(value) && value.nodeType === 11;
}
export function isString(value) {
  return typeof value === 'string';
}
