// @glimmer/runtime/dist/commonjs/es2017/lib/helpers/invoke.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.invokeHelper = invokeHelper;

var _env = require("@glimmer/env");

var _validator = require("@glimmer/validator");

var _util = require("@glimmer/util");

var _manager = require("@glimmer/manager");

var _arguments = require("../vm/arguments");

var _owner = require("@glimmer/owner");

var _destroyable = require("@glimmer/destroyable");

let ARGS_CACHES = _env.DEBUG ? new WeakMap() : undefined;

function getArgs(proxy) {
  return (0, _validator.getValue)(_env.DEBUG ? ARGS_CACHES.get(proxy) : proxy.argsCache);
}

class SimpleArgsProxy {
  constructor(context, computeArgs = () => _arguments.EMPTY_ARGS) {
    let argsCache = (0, _validator.createCache)(() => computeArgs(context));

    if (_env.DEBUG) {
      ARGS_CACHES.set(this, argsCache);
      Object.freeze(this);
    } else {
      this.argsCache = argsCache;
    }
  }

  get named() {
    return getArgs(this).named || _arguments.EMPTY_NAMED;
  }

  get positional() {
    return getArgs(this).positional || _arguments.EMPTY_POSITIONAL;
  }

} ////////////


function invokeHelper(context, definition, computeArgs) {
  if (_env.DEBUG && (typeof context !== 'object' || context === null)) {
    throw new Error(`Expected a context object to be passed as the first parameter to invokeHelper, got ${context}`);
  }

  const owner = (0, _owner.getOwner)(context);
  const internalManager = (0, _manager.getInternalHelperManager)(definition); // TODO: figure out why assert isn't using the TS assert thing

  if (_env.DEBUG && !internalManager) {
    throw new Error(`Expected a helper definition to be passed as the second parameter to invokeHelper, but no helper manager was found. The definition value that was passed was \`${(0, _util.debugToString)(definition)}\`. Did you use setHelperManager to associate a helper manager with this value?`);
  }

  if (_env.DEBUG && typeof internalManager === 'function') {
    throw new Error('Found a helper manager, but it was an internal built-in helper manager. `invokeHelper` does not support internal helpers yet.');
  }

  const manager = internalManager.getDelegateFor(owner);
  let args = new SimpleArgsProxy(context, computeArgs);
  let bucket = manager.createHelper(definition, args);
  let cache;

  if ((0, _manager.hasValue)(manager)) {
    cache = (0, _validator.createCache)(() => {
      if (_env.DEBUG && ((0, _destroyable.isDestroying)(cache) || (0, _destroyable.isDestroyed)(cache))) {
        throw new Error(`You attempted to get the value of a helper after the helper was destroyed, which is not allowed`);
      }

      return manager.getValue(bucket);
    });
    (0, _destroyable.associateDestroyableChild)(context, cache);
  } else {
    throw new Error('TODO: unreachable, to be implemented with hasScheduledEffect');
  }

  if ((0, _manager.hasDestroyable)(manager)) {
    let destroyable = manager.getDestroyable(bucket);
    (0, _destroyable.associateDestroyableChild)(cache, destroyable);
  }

  return cache;
}
