// @glimmer/runtime/dist/modules/es5/lib/vm/render-result.js
import { associateDestroyableChild, registerDestructor } from '@glimmer/destroyable';
import { clear } from '../bounds';
import UpdatingVMImpl from './update';

var RenderResultImpl = /*#__PURE__*/function () {
  function RenderResultImpl(env, updating, bounds, drop) {
    var _this = this;

    this.env = env;
    this.updating = updating;
    this.bounds = bounds;
    this.drop = drop;
    associateDestroyableChild(this, drop);
    registerDestructor(this, function () {
      return clear(_this.bounds);
    });
  }

  var _proto = RenderResultImpl.prototype;

  _proto.rerender = function rerender(_temp) {
    var _ref = _temp === void 0 ? {
      alwaysRevalidate: false
    } : _temp,
        _ref$alwaysRevalidate = _ref.alwaysRevalidate,
        alwaysRevalidate = _ref$alwaysRevalidate === void 0 ? false : _ref$alwaysRevalidate;

    var env = this.env,
        updating = this.updating;
    var vm = new UpdatingVMImpl(env, {
      alwaysRevalidate: alwaysRevalidate
    });
    vm.execute(updating, this);
  };

  _proto.parentElement = function parentElement() {
    return this.bounds.parentElement();
  };

  _proto.firstNode = function firstNode() {
    return this.bounds.firstNode();
  };

  _proto.lastNode = function lastNode() {
    return this.bounds.lastNode();
  };

  _proto.handleException = function handleException() {
    throw 'this should never happen';
  };

  return RenderResultImpl;
}();

export { RenderResultImpl as default };
