// @glimmer/runtime/dist/modules/es5/lib/vm/stack.js
import { $fp, $sp } from '@glimmer/vm';
import { REGISTERS } from '../symbols';
import { initializeRegistersWithSP } from './low-level';

var EvaluationStackImpl = /*#__PURE__*/function () {
  // fp -> sp
  function EvaluationStackImpl(stack, registers) {
    if (stack === void 0) {
      stack = [];
    }

    this.stack = stack;
    this[REGISTERS] = registers;

    if (false
    /* LOCAL_DEBUG */
    ) {
        Object.seal(this);
      }
  }

  EvaluationStackImpl.restore = function restore(snapshot) {
    return new this(snapshot.slice(), initializeRegistersWithSP(snapshot.length - 1));
  };

  var _proto = EvaluationStackImpl.prototype;

  _proto.push = function push(value) {
    this.stack[++this[REGISTERS][$sp]] = value;
  };

  _proto.dup = function dup(position) {
    if (position === void 0) {
      position = this[REGISTERS][$sp];
    }

    this.stack[++this[REGISTERS][$sp]] = this.stack[position];
  };

  _proto.copy = function copy(from, to) {
    this.stack[to] = this.stack[from];
  };

  _proto.pop = function pop(n) {
    if (n === void 0) {
      n = 1;
    }

    var top = this.stack[this[REGISTERS][$sp]];
    this[REGISTERS][$sp] -= n;
    return top;
  };

  _proto.peek = function peek(offset) {
    if (offset === void 0) {
      offset = 0;
    }

    return this.stack[this[REGISTERS][$sp] - offset];
  };

  _proto.get = function get(offset, base) {
    if (base === void 0) {
      base = this[REGISTERS][$fp];
    }

    return this.stack[base + offset];
  };

  _proto.set = function set(value, offset, base) {
    if (base === void 0) {
      base = this[REGISTERS][$fp];
    }

    this.stack[base + offset] = value;
  };

  _proto.slice = function slice(start, end) {
    return this.stack.slice(start, end);
  };

  _proto.capture = function capture(items) {
    var end = this[REGISTERS][$sp] + 1;
    var start = end - items;
    return this.stack.slice(start, end);
  };

  _proto.reset = function reset() {
    this.stack.length = 0;
  };

  _proto.toArray = function toArray() {
    return this.stack.slice(this[REGISTERS][$fp], this[REGISTERS][$sp] + 1);
  };

  return EvaluationStackImpl;
}();

export { EvaluationStackImpl as default };
