// @glimmer/runtime/dist/modules/es2017/lib/environment.js
var _a;

import { DEBUG } from '@glimmer/env';
import { assert, symbol } from '@glimmer/util';
import { track, updateTag } from '@glimmer/validator';
import { DOMChangesImpl, DOMTreeConstruction } from './dom/helper';
import { RuntimeProgramImpl } from '@glimmer/program';
import DebugRenderTree from './debug-render-tree';
export const TRANSACTION = symbol('TRANSACTION');

class TransactionImpl {
  constructor() {
    this.scheduledInstallModifiers = [];
    this.scheduledUpdateModifiers = [];
    this.createdComponents = [];
    this.updatedComponents = [];
  }

  didCreate(component) {
    this.createdComponents.push(component);
  }

  didUpdate(component) {
    this.updatedComponents.push(component);
  }

  scheduleInstallModifier(modifier) {
    this.scheduledInstallModifiers.push(modifier);
  }

  scheduleUpdateModifier(modifier) {
    this.scheduledUpdateModifiers.push(modifier);
  }

  commit() {
    let {
      createdComponents,
      updatedComponents
    } = this;

    for (let i = 0; i < createdComponents.length; i++) {
      let {
        manager,
        state
      } = createdComponents[i];
      manager.didCreate(state);
    }

    for (let i = 0; i < updatedComponents.length; i++) {
      let {
        manager,
        state
      } = updatedComponents[i];
      manager.didUpdate(state);
    }

    let {
      scheduledInstallModifiers,
      scheduledUpdateModifiers
    } = this; // Prevent a transpilation issue we guard against in Ember, the
    // throw-if-closure-required issue

    let manager, state;

    for (let i = 0; i < scheduledInstallModifiers.length; i++) {
      let modifier = scheduledInstallModifiers[i];
      manager = modifier.manager;
      state = modifier.state;
      let modifierTag = manager.getTag(state);

      if (modifierTag !== null) {
        let tag = track( // eslint-disable-next-line no-loop-func
        () => manager.install(state), DEBUG && `- While rendering:\n  (instance of a \`${modifier.definition.resolvedName || manager.getDebugName(modifier.definition.state)}\` modifier)`);
        updateTag(modifierTag, tag);
      } else {
        manager.install(state);
      }
    }

    for (let i = 0; i < scheduledUpdateModifiers.length; i++) {
      let modifier = scheduledUpdateModifiers[i];
      manager = modifier.manager;
      state = modifier.state;
      let modifierTag = manager.getTag(state);

      if (modifierTag !== null) {
        let tag = track( // eslint-disable-next-line no-loop-func
        () => manager.update(state), DEBUG && `- While rendering:\n  (instance of a \`${modifier.definition.resolvedName || manager.getDebugName(modifier.definition.state)}\` modifier)`);
        updateTag(modifierTag, tag);
      } else {
        manager.update(state);
      }
    }
  }

}

export class EnvironmentImpl {
  constructor(options, delegate) {
    this.delegate = delegate;
    this[_a] = null; // Delegate methods and values

    this.isInteractive = this.delegate.isInteractive;
    this.debugRenderTree = this.delegate.enableDebugTooling ? new DebugRenderTree() : undefined;

    if (options.appendOperations) {
      this.appendOperations = options.appendOperations;
      this.updateOperations = options.updateOperations;
    } else if (options.document) {
      this.appendOperations = new DOMTreeConstruction(options.document);
      this.updateOperations = new DOMChangesImpl(options.document);
    } else if (DEBUG) {
      throw new Error('you must pass document or appendOperations to a new runtime');
    }
  }

  getAppendOperations() {
    return this.appendOperations;
  }

  getDOM() {
    return this.updateOperations;
  }

  begin() {
    var _b;

    (false && assert(!this[TRANSACTION], 'A glimmer transaction was begun, but one already exists. You may have a nested transaction, possibly caused by an earlier runtime exception while rendering. Please check your console for the stack trace of any prior exceptions.'));
    (_b = this.debugRenderTree) === null || _b === void 0 ? void 0 : _b.begin();
    this[TRANSACTION] = new TransactionImpl();
  }

  get transaction() {
    return this[TRANSACTION];
  }

  didCreate(component) {
    this.transaction.didCreate(component);
  }

  didUpdate(component) {
    this.transaction.didUpdate(component);
  }

  scheduleInstallModifier(modifier) {
    if (this.isInteractive) {
      this.transaction.scheduleInstallModifier(modifier);
    }
  }

  scheduleUpdateModifier(modifier) {
    if (this.isInteractive) {
      this.transaction.scheduleUpdateModifier(modifier);
    }
  }

  commit() {
    var _b;

    let transaction = this.transaction;
    this[TRANSACTION] = null;
    transaction.commit();
    (_b = this.debugRenderTree) === null || _b === void 0 ? void 0 : _b.commit();
    this.delegate.onTransactionCommit();
  }

}
_a = TRANSACTION;
export function runtimeContext(options, delegate, artifacts, resolver) {
  return {
    env: new EnvironmentImpl(options, delegate),
    program: new RuntimeProgramImpl(artifacts.constants, artifacts.heap),
    resolver: resolver
  };
}
export function inTransaction(env, cb) {
  if (!env[TRANSACTION]) {
    env.begin();

    try {
      cb();
    } finally {
      env.commit();
    }
  } else {
    cb();
  }
}
export default EnvironmentImpl;
