// @glimmer/compiler/dist/commonjs/es2017/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "defaultId", {
  enumerable: true,
  get: function () {
    return _compiler.defaultId;
  }
});
Object.defineProperty(exports, "precompile", {
  enumerable: true,
  get: function () {
    return _compiler.precompile;
  }
});
Object.defineProperty(exports, "precompileJSON", {
  enumerable: true,
  get: function () {
    return _compiler.precompileJSON;
  }
});
Object.defineProperty(exports, "ProgramSymbols", {
  enumerable: true,
  get: function () {
    return _builder.ProgramSymbols;
  }
});
Object.defineProperty(exports, "buildStatement", {
  enumerable: true,
  get: function () {
    return _builder.buildStatement;
  }
});
Object.defineProperty(exports, "buildStatements", {
  enumerable: true,
  get: function () {
    return _builder.buildStatements;
  }
});
Object.defineProperty(exports, "s", {
  enumerable: true,
  get: function () {
    return _builder.s;
  }
});
Object.defineProperty(exports, "c", {
  enumerable: true,
  get: function () {
    return _builder.c;
  }
});
Object.defineProperty(exports, "unicode", {
  enumerable: true,
  get: function () {
    return _builder.unicode;
  }
});
Object.defineProperty(exports, "NEWLINE", {
  enumerable: true,
  get: function () {
    return _builder.NEWLINE;
  }
});
Object.defineProperty(exports, "WireFormatDebugger", {
  enumerable: true,
  get: function () {
    return _wireFormatDebug.default;
  }
});

var _compiler = require("./lib/compiler");

var _builder = require("./lib/builder/builder");

var _wireFormatDebug = _interopRequireDefault(require("./lib/wire-format-debug"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
