// @glimmer/wire-format/dist/amd/es5/glimmer-wire-format.js
define('@glimmer/wire-format', ['exports'], function (exports) { 'use strict';

  function is(variant) {
    return function (value) {
      return Array.isArray(value) && value[0] === variant;
    };
  } // Statements

  var isFlushElement = is(12
  /* FlushElement */
  );
  function isAttribute(val) {
    return val[0] === 14
    /* StaticAttr */
    || val[0] === 15
    /* DynamicAttr */
    || val[0] === 22
    /* TrustingDynamicAttr */
    || val[0] === 16
    /* ComponentAttr */
    || val[0] === 24
    /* StaticComponentAttr */
    || val[0] === 23
    /* TrustingComponentAttr */
    || val[0] === 17
    /* AttrSplat */
    || val[0] === 4
    /* Modifier */
    ;
  }
  function isStringLiteral(expr) {
    return typeof expr === 'string';
  }
  function getStringFromValue(expr) {
    return expr;
  }
  function isArgument(val) {
    return val[0] === 21
    /* StaticArg */
    || val[0] === 20
    /* DynamicArg */
    ;
  }
  function isHelper(expr) {
    return Array.isArray(expr) && expr[0] === 28
    /* Call */
    ;
  } // Expressions

  var isGet = is(30
  /* GetSymbol */
  );

  exports.getStringFromValue = getStringFromValue;
  exports.is = is;
  exports.isArgument = isArgument;
  exports.isAttribute = isAttribute;
  exports.isFlushElement = isFlushElement;
  exports.isGet = isGet;
  exports.isHelper = isHelper;
  exports.isStringLiteral = isStringLiteral;

  Object.defineProperty(exports, '__esModule', { value: true });

});
