// @glimmer/runtime/dist/commonjs/es5/lib/compiled/opcodes/debugger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setDebuggerCallback = setDebuggerCallback;
exports.resetDebuggerCallback = resetDebuggerCallback;

var _reference = require("@glimmer/reference");

var _util = require("@glimmer/util");

var _opcodes = require("../../opcodes");

var _symbols2 = require("../../symbols");

function debugCallback(context, get) {
  // eslint-disable-next-line no-console
  console.info('Use `context`, and `get(<path>)` to debug this template.'); // for example...
  // eslint-disable-next-line no-unused-expressions

  context === get('this'); // eslint-disable-next-line no-debugger

  debugger;
}

var callback = debugCallback; // For testing purposes

function setDebuggerCallback(cb) {
  callback = cb;
}

function resetDebuggerCallback() {
  callback = debugCallback;
}

var ScopeInspector = /*#__PURE__*/function () {
  function ScopeInspector(scope, symbols, evalInfo) {
    this.scope = scope;
    this.locals = (0, _util.dict)();

    for (var i = 0; i < evalInfo.length; i++) {
      var slot = evalInfo[i];
      var name = symbols[slot - 1];
      var ref = scope.getSymbol(slot);
      this.locals[name] = ref;
    }
  }

  var _proto = ScopeInspector.prototype;

  _proto.get = function get(path) {
    var scope = this.scope,
        locals = this.locals;
    var parts = path.split('.');

    var _path$split = path.split('.'),
        head = _path$split[0],
        tail = _path$split.slice(1);

    var evalScope = scope.getEvalScope();
    var ref;

    if (head === 'this') {
      ref = scope.getSelf();
    } else if (locals[head]) {
      ref = locals[head];
    } else if (head.indexOf('@') === 0 && evalScope[head]) {
      ref = evalScope[head];
    } else {
      ref = this.scope.getSelf();
      tail = parts;
    }

    return tail.reduce(function (r, part) {
      return (0, _reference.childRefFor)(r, part);
    }, ref);
  };

  return ScopeInspector;
}();

_opcodes.APPEND_OPCODES.add(103
/* Debugger */
, function (vm, _ref) {
  var _symbols = _ref.op1,
      _evalInfo = _ref.op2;

  var symbols = vm[_symbols2.CONSTANTS].getArray(_symbols);

  var evalInfo = vm[_symbols2.CONSTANTS].getArray((0, _util.decodeHandle)(_evalInfo));

  var inspector = new ScopeInspector(vm.scope(), symbols, evalInfo);
  callback((0, _reference.valueForRef)(vm.getSelf()), function (path) {
    return (0, _reference.valueForRef)(inspector.get(path));
  });
});
