// @glimmer/low-level/dist/commonjs/es5/lib/asm/stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Stack = void 0;

var Stack = /*#__PURE__*/function () {
  function Stack(vec) {
    if (vec === void 0) {
      vec = [];
    }

    this.vec = vec;
  }

  var _proto = Stack.prototype;

  _proto.clone = function clone() {
    return new Stack(this.vec.slice());
  };

  _proto.sliceFrom = function sliceFrom(start) {
    return new Stack(this.vec.slice(start));
  };

  _proto.slice = function slice(start, end) {
    return new Stack(this.vec.slice(start, end));
  };

  _proto.copy = function copy(from, to) {
    this.vec[to] = this.vec[from];
  } // TODO: how to model u64 argument?
  ;

  _proto.writeRaw = function writeRaw(pos, value) {
    // TODO: Grow?
    this.vec[pos] = value;
  } // TODO: partially decoded enum?
  ;

  _proto.getRaw = function getRaw(pos) {
    return this.vec[pos];
  };

  _proto.reset = function reset() {
    this.vec.length = 0;
  };

  _proto.len = function len() {
    return this.vec.length;
  };

  return Stack;
}();

exports.Stack = Stack;
