// @glimmer/compiler/dist/commonjs/es2017/lib/shared/list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OptionalList = OptionalList;
exports.EmptyList = exports.PresentList = void 0;

var _util = require("@glimmer/util");

class PresentList {
  constructor(list) {
    this.list = list;
  }

  toArray() {
    return this.list;
  }

  map(callback) {
    let result = (0, _util.mapPresent)(this.list, callback);
    return new PresentList(result);
  }

  filter(predicate) {
    let out = [];

    for (let item of this.list) {
      if (predicate(item)) {
        out.push(item);
      }
    }

    return OptionalList(out);
  }

  toPresentArray() {
    return this.list;
  }

  into({
    ifPresent
  }) {
    return ifPresent(this);
  }

}

exports.PresentList = PresentList;

class EmptyList {
  constructor() {
    this.list = [];
  }

  map(_callback) {
    return new EmptyList();
  }

  filter(_predicate) {
    return new EmptyList();
  }

  toArray() {
    return this.list;
  }

  toPresentArray() {
    return null;
  }

  into({
    ifEmpty
  }) {
    return ifEmpty();
  }

} // export type OptionalList<T> = PresentList<T> | EmptyList<T>;


exports.EmptyList = EmptyList;

function OptionalList(value) {
  if ((0, _util.isPresent)(value)) {
    return new PresentList(value);
  } else {
    return new EmptyList();
  }
}
